{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `cloudwatcher` Python API usage\n",
    "\n",
    "The purpose of this page is to provide a quick overview of the `cloudwatcher` Python API. The package consists of two user-facing classes:\n",
    "\n",
    "- `MetricWatcher`: This class is used to interact with AWS CloudWatch metrics.\n",
    "- `LogWatcher`: This class is used to interact with AWS CloudWatch logs.\n",
    "\n",
    "Both of these classes inherit from the `CloudWatcher` class.\n",
    "\n",
    "## `MetricWatcher`: convenient interface to AWS CloudWatch metrics\n",
    "\n",
    "`MetricWatcher` can be used to interact with AWS CloudWatch metrics. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `MetricWatcher` initialization\n",
    "\n",
    "As described in the Login credentials section, the AWS credentials can be sourced from environment variables: `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`.\n",
    "Alternatively, you can pass the values as arguments to the `MetricWatcher` constructor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cloudwatcher.metricwatcher import MetricWatcher\n",
    "from cloudwatcher.preset import Dimension\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "instance_id = os.environ.get(\"INSTANCE_ID\")\n",
    "mw = MetricWatcher(\n",
    "    namespace=\"NepheleNamespaceEC2\",\n",
    "    metric_name=\"mem_used\",\n",
    "    metric_id=\"mem_used\",\n",
    "    metric_unit=\"Bytes\",\n",
    "    dimensions_list=[Dimension(Name=\"InstanceId\", Value=instance_id)],\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `MetricWatcher` presets\n",
    "\n",
    "As you can see there are multiple arguments that can be passed to `MetricWatcher` constructor. In order to improve the UX when using `MetricWatcher` `cloudwatcher` package provides a few presets that can be used to query the data reported by `CloudWatchAgent` within certain systems. Additionally, custom presets can be defined by the user and used in the same way.\n",
    "\n",
    "Presets are JSON-formatted files that provide parameter bundles for `MetricWatcher` initialization.\n",
    "\n",
    "#### Usage\n",
    "\n",
    "Listing available presets:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-style: italic\">                 Presets available in: /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets                  </span>\n",
       "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓\n",
       "┃<span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\"> Name                                   </span>┃<span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\"> Path                                                                   </span>┃\n",
       "┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩\n",
       "│ nephele_disk_used_percent              │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_di… </span>│\n",
       "│ nephele_mem_cached                     │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_me… </span>│\n",
       "│ nephele_mem                            │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_me… </span>│\n",
       "│ nephele_disk_used_percent_nephele_data │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_di… </span>│\n",
       "│ nephele_cpu_usage_user                 │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_cp… </span>│\n",
       "│ nephele_processes_dead                 │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_pr… </span>│\n",
       "│ nephele_swap_used_percent              │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_sw… </span>│\n",
       "│ nephele_swap_used                      │<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_sw… </span>│\n",
       "└────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[3m                 Presets available in: /Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets                  \u001b[0m\n",
       "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓\n",
       "┃\u001b[1;35m \u001b[0m\u001b[1;35mName                                  \u001b[0m\u001b[1;35m \u001b[0m┃\u001b[1;35m \u001b[0m\u001b[1;35mPath                                                                  \u001b[0m\u001b[1;35m \u001b[0m┃\n",
       "┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩\n",
       "│ nephele_disk_used_percent              │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_di…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_mem_cached                     │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_me…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_mem                            │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_me…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_disk_used_percent_nephele_data │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_di…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_cpu_usage_user                 │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_cp…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_processes_dead                 │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_pr…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_swap_used_percent              │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_sw…\u001b[0m\u001b[2m \u001b[0m│\n",
       "│ nephele_swap_used                      │\u001b[2m \u001b[0m\u001b[2m/Users/stolarczykmj/code/cloudwatcher/cloudwatcher/presets/nephele_sw…\u001b[0m\u001b[2m \u001b[0m│\n",
       "└────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from cloudwatcher.preset import PresetFilesInventory\n",
    "from rich.console import Console\n",
    "\n",
    "pfi = PresetFilesInventory()\n",
    "Console().print(pfi.presets_table)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using a preset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'MetricDataResults': [{'Id': 'nephele', 'Label': 'mem_used', 'Timestamps': [datetime.datetime(2023, 7, 19, 16, 45, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 44, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 43, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 42, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 41, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 40, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 39, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 38, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 37, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 36, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 35, tzinfo=tzutc()), datetime.datetime(2023, 7, 19, 16, 34, tzinfo=tzutc())], 'Values': [486064128.0, 485814272.0, 4685066240.0, 6207594496.0, 4992217088.0, 4720185344.0, 2435854336.0, 2444738560.0, 2400739328.0, 11007488000.0, 2191474688.0, 576376832.0], 'StatusCode': 'Complete'}], 'Messages': [], 'ResponseMetadata': {'RequestId': 'bbc8d20e-7879-447e-87e5-019b3769220f', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': 'bbc8d20e-7879-447e-87e5-019b3769220f', 'content-type': 'text/xml', 'content-length': '1596', 'date': 'Wed, 19 Jul 2023 18:35:02 GMT'}, 'RetryAttempts': 0}}\n"
     ]
    }
   ],
   "source": [
    "from cloudwatcher.preset import MetricWatcherSetup\n",
    "mw_setup = MetricWatcherSetup.from_json(pfi.get_preset_path(\"nephele_mem\"))\n",
    "mw_setup.upsert_dimensions([f\"InstanceId:{instance_id}\"])\n",
    "mw = MetricWatcher(**mw_setup.to_dict())\n",
    "query_kwargs = {\n",
    "    \"days\": 5,\n",
    "    \"hours\": 0,\n",
    "    \"minutes\": 0,\n",
    "    \"stat\": \"Maximum\",\n",
    "    \"period\": 60,\n",
    "}\n",
    "response = mw.query_ec2_metrics(**query_kwargs)\n",
    "print(response)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "### Querying AWS CloudWatch metrics\n",
    "\n",
    "In order to specify the EC2 instace query settings (period, granularity, etc.), the user would need to provide multiple parameters. To make it easier, there are a few sensible presets that can be used to select the query settings, which are passed to `query_ec2_metrics` method. These presets are defined to query the data reported by `CloudWatchAgent` within the last day, hour or minute."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The presets can be used by passing the `query_preset` argument to the functions presented below. Alternatively, users can pass `query_kwargs` argument, which overrides the preset values.\n",
    "\n",
    "### Logging methods\n",
    "\n",
    "There is a method that can be used to log the metric to the screen. The EC2 instance is automatically queried if the query response is not provided."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File saving methods\n",
    "\n",
    "There are number of methods that can be used to save the metric data to a file. Again, the EC2 instance is automatically queried if the query response is not provided.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGjCAYAAADZ3JQcAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAABR/0lEQVR4nO3deVyUdeIH8M8zAwww3MyAIiADHoD3nWWbJpalZbWWtVZ2W2lmtrWZHdtpuf3MPFI7tcOyu90OM0vLyspbVDyZUbyAGZRjBgZm5vv7AyFHQBmcmWeOz/v14sXOwxwfwmU+PM/3kIQQAkRERER+SiF3ACIiIqJzwTJDREREfo1lhoiIiPwaywwRERH5NZYZIiIi8mssM0REROTXWGaIiIjIr7HMEBERkV9jmSEiIiK/xjJDRER0DmbNmgWFQgGz2Sx3FI9LSkrC/fffL3eMJlhmiIiIzsGOHTuQmZkJtVotdxS32LJlC/7973+jsrLS6fixY8dQWlqKnj17ypSsZRL3ZiIiImq7/v37Iy0tDZ9//rncUdxi+vTpmD9/PsrLy6FQ/HXOQwgBq9WKsLAwp+O+wLfSEBER+REhBAoKCtCjRw+5o7Sa3W6HzWZr8esbN25Er169mhQWSZIQHh7uc0UGYJkhIgoKI0aMwODBg7Fu3ToMHToUarUanTp1wjfffAMA+Oabb3DeeedBrVajd+/e2LhxY5PnKCwsxK233or27dtDrVZj4MCBWLVqldN9fv31V0iShP/+97+YMWMGUlNTER0djRtuuAE1NTWoqKjA5MmTkZycjMTEREydOhWuXiAYMGAARo8e3eT42LFj0a1bN6dj33zzDYYOHQqtVovIyEh07twZDz74oMvfFwAUFxfj9ttvR2JiIuLj4/HPf/4Te/fuhcViaVWZKSgowLhx46DVahEREYGhQ4di8+bNjV9///33IUkSfvnllyaPnTdvHpRKJXbs2OFS7oafx5dffolHHnkEqampCAkJQWlpaZPX2LlzJyRJwvfff9/4OEmSGsfI3HnnnejQoUPj/e12OyIiIjB9+nQsWbIE3bt3R0REBC644ALs27cPQgi8/PLL6Ny5M6KiojB69GgYjcYmr7tmzRpcfvnliIuLQ0JCAkaPHo39+/ef9b+nE0FERAEvKSlJ5OTkiLS0NPHvf/9bzJkzRyQlJYnY2FixaNEikZWVJV544QXx/PPPi6ioKNGzZ0+nx2/dulXExcWJnJwc8eKLL4r58+eLwYMHi5CQELFt27bG+7366qsCgOjVq5e46qqrxIIFC8T1118vAIjHHntM9OzZU9x8881i4cKF4pJLLhEAxGeffdbq78Nms4mIiAjx6KOPNvlap06dxPXXX994+6233hIAxNixY8WCBQvEggULxN133y2uuuoql78vo9EoMjMzRbt27cQzzzwj5s6dK3JycsTgwYMFAFFQUHDG3D///LOIjIwUvXv3FnPmzBGzZ88WOp1OJCYmCqPRKIQQYvPmzQKAeP31150ea7FYRPv27cU//vEPl3M3/Dxyc3PFsGHDxNy5c8Xzzz/fbMbDhw+LuXPnCgDi9ttvF++++6549913xc6dO4UQQgwcOFCMHDmy8f47d+5s/Fn36dNHzJ49Wzz66KNCqVSKyy67TNxyyy3ioosuEnPnzhWTJk0SAMSUKVOcXvPtt98WCoVCXHrppWLevHnixRdfFKmpqSIjI0NYLJYz/jc9FcsMEVGAKy4uFgBEcnKyOHLkSOPxhjeu7OxsUV5e3nh82rRpQpIkUVNTI4QQoqamRmRmZoqRI0eK2traxvtZLBaRkpIi7rvvvsZj99xzjwAgHn/88cZjNptNaLVaAUAsXry48Xh5ebmQJKnZYtKSgoICAUAsX77c6XhlZaWQJEk899xzjce6d+8ubrjhhhafy5Xva+zYsUKr1YpDhw41HisqKhKhoaFCpVIJm83W4uuUlZUJrVYr8vLyhNVqbTy+Y8cOAUDMmjVLCCFEdXW1UCgU4sEHH3R6/EsvvSRCQkLE3r17Xc7d8PO44447hMPhaDFjgx9//FEAEGvWrHE67nA4hFqtFg8//HDjseXLlwsAYvjw4U45rrnmGgFAjB8/3uk1+/btK84///zG2/n5+SIsLEw888wzTq+1detWAUB8+umnZ83bIMS18zhERORvtm3bBgD497//jfbt2zcej4qKAgD85z//QUxMTOPx2NhYKBSKxrERb775Jg4cOIBPP/0U5eXlTs/duXNnFBYWOr1Whw4d8OSTTzYeUyqViIiIQL9+/XDXXXc1Hler1VAqlQgJaf1bUcP30qtXrybHhRBOl3tOnDiB/fv3w2QyITExsclztfb72r59Oz755BO88MILTpdZUlNTkZWVhfDwcCiVyhYzL168GKWlpZg3bx7CwsIaj+fm5iIuLg67du0CAISHhyMrK6vxNgCYzWa8+OKLmDBhAjp16uRS7ob/LomJiXj55ZchSVKLGU+9P4AmM5b2798Ps9ns9N+34b7z589HaGho4/GoqCiEhoZi7ty5Tq8ZGxsLu93eePvpp59Geno6Jk6c6HT5KSUlBaGhoU7fx9mwzBARBbj8/HwAwJVXXul0fPfu3YiIiMCIESOcju/ZswdZWVmNb1Cffvop7HY7+vTp0+zzX3/99Y3/e/v27bjuuuuc3tzNZjMOHTqE22+/3elxer0eNpsN2dnZrf5etm7dioiICHTu3Nnp+JYtWwDA6c324YcfxtSpU5GamopRo0bhhhtuwJgxYxrLU2u/r88//xwKhcKpiDWw2WxnHS/z3//+FwMHDmzyfQohUF1d7VRwunXr1vjzAoC5c+eivLwcTzzxROMxV38eV155ZWNxPZtt27YhNTUV8fHxTscbMp1acvLz89G5c+cm39fu3bsxZMgQJCQkOB3fs2cPRo0aBQCwWq34+uuvYbFYkJSU1GyW6OjoVmUGWGaIiALetm3b0L59e6SkpDgd37p1K7p37w6VStXk+KlvWtu3b8fYsWMxceLEZp8/PT0dAHDgwAGUl5ejf//+Tl/fvn07HA5Hk+Nbt24F0PQswJnk5+eje/fuTWbUrF+/HjExMejYsWPjsfvuuw8jR47ERx99hE8//RRjx47FwIEDsXbtWoSFhbX6+8rPz4dOp2vyBl9WVobCwsJmS86pCgoKcMUVVzQ5XlRUBKvVitzc3MZjubm5+N///ger1Qqr1YqXXnoJd911V2MWwPWfx5AhQ86Y71Tbtm1rctar4XhoaKhTcdm2bRsGDx7sdD+Hw4H8/HxMmjTJ6bjJZMLhw4cbf9aFhYWwWCx45plncN555zWbpbkcLWGZISIKcC29QW3durXxL+UGdXV12L17N6677rrGY5WVlWjXrh3y8vLO+jpA0zehhtJy+vH8/HyoVCp07dq11d9LQUEBzj//fKdjVqsVX331Fbp3797kUkrnzp0xY8YMzJgxA08++SSefvpp7NixA3369Gn192WxWJo9vnjxYjgcjrOemZEkqUlhBOovFwFw+hl069YNdrsde/fuxWeffYbq6mrMmDHD6XHn+vNoicPhwI4dO5qcqWt4rq5duzaeRaqoqMCBAwdw9913O91v3759sFgszf6sgb+Ka8OCfDk5OWf9PlqDU7OJiAKY3W7Hzp07m7y5GI1GHD16tMnxgoIC1NXVOZ0t0el0+Oabb1BTU9Pk+U+d4pufnw+FQtHkzb1h3Map400ajufk5Lg0ZqampsapXAgh8M9//hNGo9HpdZubAhwaGgpJkhrHDbX2++rYsSMMBgP0en3jsaKiIrz88ssAcNYyk52djXXr1jlNQd+zZw/mzp2L6667DpmZmY3HG6aWr1u3DrNnz8bkyZPRrl07p+c7159HS0pLS1FdXY3U1NQmX8vPz29ytk4I0ezYJaD5MU3AX2UmIyMDkiTh008/bfJaNpsNx48fb1XmBjwzQ0QUwPbu3YuamppWny1pbgDo1KlTcdddd2HAgAGYMGECoqOjYTAY8N133+Gaa67BY4891vjYTp06ITIysslrtXTp4oILLnDp++nXrx8+//xzPPjgg2jfvj2++uor1NXVAfirVFgslsZxMgMHDkRCQgL+/PNPLFmyBPfee29jOWjt93XjjTfi1VdfxciRIzF58mRUVFRg3rx5AID4+PgmJe1006ZNw7XXXosxY8Zg9OjRKCoqwvz585GamorFixc73Tc7OxtKpRIzZsyAw+HAv/71rybPd64/j5ZoNBpERUXh9ddfh0qlgkqlwujRoxEWFob9+/fjtttua7zvmc7CqVSqJuNotm3bhvT0dMTGxgKo3+PphhtuwLJly1BRUYHLLrsMdrsd+/btw2effYYPP/zQpctjnJpNRBTAPvroIwFAbN++3en47NmzBQBx4sQJp+MPP/ywiImJaTKN9+OPPxaDBg0SsbGxIjo6WuTk5IhJkyaJ/fv3N94nNzdXjB07tkmG2NhYMXXqVKdjVVVVQpIk8dJLL7n0/ej1ejF06FAREREhMjMzxcyZM8V3330nAIiffvpJCCHE8ePHxZQpU0T37t1FVFSUiImJEQMGDBBLliwRdrvd5e9LiPo1a7KysoRKpRLdu3cXr732msjLyxMXXnhhq3LPnz9f6HQ6ERYWJnQ6nXj44YdFVVVVs/ft0qVLk+ntpzuXn8eZfPbZZ6JLly4iJCRESJIkysvLxZ9//ikAiK+//rrxfvfee6/QaDRNHn/llVeK3r17Nzk+cOBAMXr0aKdjNTU14vnnnxe5ubkiIiJCJCYmigEDBognn3xSmM1ml3JzbyYiIiLyaxwzQ0RERH6NY2aIiEh21dXVTRaAO11iYqLT4mxEDVhmiIhIdkuXLsU999xzxvusW7euxTVJKLhxzAwREcnu8OHDKCgoOON9Bg4c6LTtAlEDlhkiIiLyaxwATERERH4t4MfMOBwOHDlyBNHR0a3aMZSIiIjkJ4RAZWUlUlJSmuzFdbqALzNHjhxBWlqa3DGIiIioDYqKiprdYuFUAV9mGrYQLyoq4sAxIiIiP1FRUYG0tLTG9/EzCfgy03BpKSYmhmWGiIjIz7RmiAgHABMREZFfY5khIiIiv8YyQ0RERH6NZYaIiIj8GssMERER+TWWGSIiIvJrLDNERETk11hmiIiIyK+xzBAREZFfY5khIiIiv8YyQ0RERH6NZYaIiIj8GssMERER+TWWGZLdO+sMqLU55I5BRER+imWGZFVuqcMTX+5AwdEKuaMQEZGfYpkhWelNZgBAobFK5iREROSvWGZIVvqTJUZfapY5CRER+SuWGZKV3mgBABQaWWaIiKhtZC0zP//8M6644gqkpKRAkiR88cUXTl8XQuCJJ55A+/btERERgby8POzdu1eesOQRBqMZPTrEQs8yQ0REbSRrmTGbzejVqxcWLFjQ7NdnzZqFuXPnYtGiRfjjjz+gVqtx6aWXoqamxstJyVMMJjMuzk6C3miGEELuOERE5IdkLTOXXXYZnn32WVx99dVNviaEwJw5c/DYY49hzJgx6NmzJ9555x0cOXKkyRkc8k9CCOhLzRjaVYvqOjtKKq1yRyIiIj/ks2Nm9Ho9jh07hry8vMZjsbGxGDRoENatWydjMnIXk7kWlVYburaLRoe4COwv5YwmIiJync+WmWPHjgEAkpOTnY4nJyc3fq05VqsVFRUVTh/kmwxGM5JjVIgMC4FOo+a4GSIiahOfLTNtNXPmTMTGxjZ+pKWlyR2JWqA3mqHTqAEAmRo1p2cTEVGb+GyZadeuHQCguLjY6XhxcXHj15ozffp0lJeXN34UFRV5NCe13allhmdmiIiorXy2zOh0OrRr1w4//PBD47GKigr88ccfGDx4cIuPU6lUiImJcfog32QwmZGRePLMjDaKZYaIiNokRM4Xr6qqwr59+xpv6/V6bNmyBQkJCUhPT8fUqVPx7LPPonPnztDpdHj88ceRkpKCq666Sr7Q5DZ6owVX9e4AoP7MzMEyC+rsDoQqfbZjExGRD5K1zGzYsAHDhg1rvD1t2jQAwIQJE7BkyRI8/PDDMJvNuOuuu3DixAkMGTIEK1asQHh4uFyRyU2EEDCccpkpJS4CCoWEojILMrVRMqcjIiJ/ImuZGTp06BkXSpMkCU8//TSefvppL6YibyiusKLGZkdaQiQAQKmQkJEYCb3RzDJDREQu4fl8koXeaEZKbATCQ5WNxzgImIiI2oJlhmRhMJmRqVU7HdNporjhJBERuYxlhmShN/41k6lBppZrzRARketYZkgWeqMZGZrTyoxGjUIjtzQgIiLXsMyQLOpnMkU6HdNp1CiusMJstcmUioiI/BHLDHmd3SFwwGSBTuM8aylBHYaY8BAOAiYiIpewzJDXHTlRDbsQSI2PcDouSRJ0XAmYiIhcxDJDXmcwmZEWH9HsSr9ZnJ5NREQuYpkhrzt15d/T6TRqFJZyEDAREbUeywx5XWEzM5ka6LQ8M0NERK5hmSGvO+uZGaP5jNtcEBERnYplhrzOYLI0WTCvQUaiGpU1NpjMtV5ORURE/oplhryqzu5AUZmlxTMzalUI2sWE81ITERG1GssMedWh49VQSBJS4iJavA+3NSAiIlewzJBXGYxmpCdGQqmQWryPTqPGfm5rQERErcQyQ17V3AaTp9NpeGaGiIhaj2WGvEpvNCNTe+Yyk8np2URE5AKWGfIqg6k1Z2aicMBkgd3B6dlERHR2LDPkVXqjGRmn7ZZ9urT4CDiEwJET1V5KRURE/oxlhrzGarPj8IlqZJ62W/bpQpQKpCdGYj+3NSAiolZgmSGvOWiyIDxEieQY1Vnvm8kNJ4mIqJVYZshr9EYzOiZGQpJanpbdQMcyQ0RErcQyQ15jMLW8J9PpdJoolhkiImoVlhnyGv0ZNpg8nU6jRiHXmiEiolZgmSGvqZ/J1Loyk6VV40h5NWrq7B5ORURE/o5lhrzGYGx5g8nTaaNViAxVwmDi2RkiIjozlhnyCkutDccqalpdZiRJgo4bThIRUSuwzJBXGIwWRKtCkKgOa/VjdJooFHIQMBERnQXLDHmFwVQ/XqY107IbcHo2ERG1BssMeYUrg38bZHHDSSIiagWWGfIKgwvTshvUT8/mlgZERHRmLDPkFfVrzJx5g8nTZWjUOG6pw3FzrYdSERFRIGCZIa8wmMzISHTtzExMeCg0USroOT2biIjOgGWGPK6ipg7GqlqXLzMBJzec5PRsIiI6A5YZ8jiD0Yz4yFDERbZ+WnYDzmgiIqKzYZkhj2vLTKYGmZzRREREZ8EyQx5nMFqgc3G8TAOdRo39nNFERERnwDJDHmcwuT4tu0GmVg2DyQyHQ7g5FRERBQqWGfK4wnO4zJSWEIlamwPHKmrcnIqIiAIFywx5XFsWzGugClEiNT6S42aIiKhFLDPkUcfNtSivrmvzmRmg/lITN5wkIqKWsMyQR+lNZmijVYhShbT5OXRca4aIiM6AZYY8Sl9qbvNMpgaZGjUKjZzRREREzWOZIY8ymMzIcHFPptPpNFEcM0NERC1imSGPOpcF8xrotGoUlVlQa3O4KRUREQUSlhnyKL3RjMxzLDPtY8IRFqLAwTKLm1IREVEgYZkhjxFCwOCGMzMKhYSMRG5rQEREzWOZIY8prbLCXGtHx4RzKzPAyenZ3NaAiIiawTJDHmMwWpASG46IMOU5Pxd3zyYiopawzJDH6I1V53yJqYFOE8WF84iIqFksM+QxeqPFjWWGZ2aIiKh5LDPkMQbjuS+Y1yBLq0ZppRWVNXVueT4iIgocLDPkMfpz2GDydHGRYYiPDIXByOnZRETkjGWGPMLhECdX/3VPmQHqLzVxWwMiIjodywx5xLGKGtTZHUhPOLetDE6l00ShkBtOEhHRaVhmyCMMRjM6xEcgLMR9/8QytRwETERETbHMkEcUGs3QaaLc+pyc0URERM1hmSGPqJ/J5L5LTMBfZ2aEEG59XiIi8m8+XWbsdjsef/xx6HQ6REREICsrC8888wzfzPyAuwf/AkBGohpVVhtKK61ufV4iIvJvIXIHOJMXX3wRCxcuxNKlS9GtWzds2LABt956K2JjYzFlyhS549EZ6I1m3HheR7c+Z3ioEh3iIlBoNCMpJtytz01ERP7Lp8vMb7/9hjFjxmDUqFEAgIyMDHzwwQf4888/ZU5GZ2KzO3CwzOK2NWZO1TBu5rzMRLc/NxER+Sefvsx0/vnn44cffsCePXsAAFu3bsUvv/yCyy67rMXHWK1WVFRUOH2Qdx05UQMhgA5xEW5/bg4CJiKi0/n0mZlHHnkEFRUVyM7OhlKphN1ux3PPPYfx48e3+JiZM2fiqaee8mJKOp3eZEZ6QiRClO7vyplaNX7dZ3L78xIRkf/y6TMzH330Ed5//30sW7YMmzZtwtKlS/HSSy9h6dKlLT5m+vTpKC8vb/woKiryYmICAH1plUcuMQENZ2a4CjAREf3Fp8/MPPTQQ3jkkUdw/fXXAwB69OiBAwcOYObMmZgwYUKzj1GpVFCpVN6MSacxmNy3W/bpMjVROFhmgc3u8MiZHyIi8j8+/W5gsVigUDhHVCqVcDgcMiWi1tAb3T8tu0GH+PpxOIeOV3vk+YmIyP/49JmZK664As899xzS09PRrVs3bN68GbNnz8Ztt90mdzQ6A4PJjEwPlRmlQkLHRLVHCxMREfkXny4z8+bNw+OPP457770XJSUlSElJwcSJE/HEE0/IHY1aUGtzoKjMc5eZgIbds80Y5rFXICIif+LTZSY6Ohpz5szBnDlz5I5CrVR03IJQpQLtPbioXf22BhwETERE9Xx6zAz5H4PRjI6JkVAoJI+9RqZGjcJSrjVDRET1WGbIrfRGs8emZTfQaaK4cB4RETVimSG38sbAXJ1GjaPlNbDU2jz6OkRE5B9YZsitDCYzdImeLTOaqDBEq0JgMFo8+jpEROQfWGbIrQxGz85kAgBJkk4OAualJiIiYpkhN6qps+PwiWqPrTFzKm5rQEREDVhmyG0OmCxQhymhjfb8dhI6TRRnNBEREQCWGXIjvdGMjolqSJLnpmU30GnrF84jIiJimSG3MZjM0Gm9s8VA/VozVRBCeOX1iIjId7HMkNvoSz0/k6lBhkaNihobjlvqvPJ6RETku1hmyG30Ju9t/hilCkFyjIqDgImIiGWG3MdgNEOnifTa6+m4rQEREYFlhtykympDSaUVOk2U115Tp4niIGAiImKZIfcwGM2ICQ9BfGSo114zU6OGnmdmiIiCHssMuYXBVL/BpDemZTeoXziPZYaIKNixzJBbGLywweTpMrVq6E1mOBycnk1EFMxYZsgtCo31Z2a8KS0hEnaHwJHyaq++LhER+RaWGXILgwxlJlSpQHpCJGc0EREFOZYZcguDyYIMLy2YdyqOmyEiIpYZOmflljqUmWu9PmYGYJkhIiKWGXIDvcmMRHUYYiO8Ny27gU7DDSeJiIIdywydMzlmMjXI1Kq5pQERUZBjmaFzpjeaZRkvAwCZmigcOl4Nq80uy+sTEZH8WGbonOmNZmRq5SkzyTEqhIcoccBkkeX1iYhIfiwzdM4MJvnOzEiSxA0niYiCHMsMnRMhRP1lJi/uln06nZYzmoiIghnLDJ2TMnMtKmtssp2ZAYAsDQcBExEFM5YZOid6oxnJMSqoVSGyZeCZGSKi4MYyQ+dEzplMDXSaKI6ZISIKYiwzdE4MJu/vyXQ6XaIaJnMtyi11suYgIiJ5tKnM1NXVoaioCLt370ZZWZm7M5Ef0cuwweTpYiNDkagOg97EszNERMGo1WWmsrISCxcuxEUXXYSYmBhkZGQgJycHWq0WHTt2xJ133on169d7Miv5IL3RItvqv6fScRAwEVHQalWZmT17NjIyMvD2228jLy8PX3zxBbZs2YI9e/Zg3bp1ePLJJ2Gz2XDJJZdg5MiR2Lt3r6dzkw8QQuCAD1xmAk5ua8BxM0REQalVU1DWr1+Pn3/+Gd26dWv26wMHDsRtt92GRYsW4e2338batWvRuXNntwYl31NSaUV1nR3pCfKtMdNAp4nCjiPlcscgIiIZtKrMfPDBB616MpVKhbvvvvucApH/KCw1IyU2AuGhSrmjQKdR439bj8gdg4iIZMDZTNRmvjCTqUHmybVmhBByRyEiIi9zqcysXr0a//d//4dff/0VALB48WKkp6dDq9XizjvvRHV1tUdCkm8yyLyNwanSEyJRY7OjuMIqdxQiIvKyVpeZ119/HSNGjMCiRYswfPhwzJw5Ew8++CBGjRqF6667Dh999BGeeuopT2YlH1M/LTtK7hgAgPBQJVLjI1DIGU1EREGn1WXmlVdewcsvv4y9e/fiiy++wBNPPIEFCxZg4cKFWLBgAd544w188sknnsxKPqa+zPjGmRmgfhAwtzUgIgo+rS4zhYWFuPLKKwEAI0eOhCRJGDhwYOPXBw0ahKKiIvcnJJ/kcAgcKLPIvpXBqTI1am5rQEQUhFpdZmpqahAREdF4W6VSQaVSOd222WzuTUc+60h5NewOgTQfmJbdoH7hPJYZIqJg0+oyI0kSKisrUVFRgfLyckiShKqqKlRUVDR+UPDQG81Ii49AqNJ3JsSxzBARBadWrTMD1K/22qVLF6fbffr0cbotSZJ705HPqp/J5DuXmID6MnOwzII6u8OnShYREXlWq8vM6tWrPZmD/Ize6FvjZQCgQ1wElAoJRWUWZGp9Y5YVERF5XqvLzEUXXeTJHORnDCYzLuqilTuGE4VCgi6x/lITywwRUfBodZk53Y4dO2C32xtvK5XKFvduosCjN5pxy/kZcsdoQndyRtPwHLmTEBGRt7R6YMHatWsxYMCAxtvnnXce+vTpg969e6N3797o2bMnVq1a5ZGQ5FtsdgeKyiw+s5XBqXRaNQo5CJiIKKi0usy8+uqruOmmm5yOrV69Gnq9HoWFhbj//vuxcOFCtwck33PoeDUUkoSUuIiz39nL6mc0cRVgIqJg0uoys2HDBlx88cVOx1JTU9GxY0dkZGTgpptuwrp169wekHyP3mhGemIklArfm72WpeX0bCKiYNPqMnPo0CHExsY23l66dCnatWvXeDshIQEmk8m96cgn6Y1mn5vJ1ECniUJxhRVmKxdwJCIKFq0uM9HR0di/f3/j7WuuuQaRkX+t/qrX6xETE+PedOSTDCbf2pPpVPGRoYiNCOXZGSKiINLqMjNo0CC88847LX59yZIlGDRokFtCkW/T++CCeQ0kSaqf0cQyQ0QUNFo9NXvatGnIy8tDYmIiHnroISQlJQEASkpK8OKLL+K9997DypUrPRaUfEf9btm+WWaA+g0n9dxwkogoaLS6zAwbNgzz5s3DAw88gNmzZyMmJgaSJKG8vBwhISGYM2dOkwHCFHisNjuOnKj26TKj06ixv5QzmoiIgoVLi+bde++9uOKKK/DJJ59g7969AIDOnTtj7NixSEtL80hA8i1FZRaEhSiQHB0ud5QWZWqjsKqgWO4YRETkJS6vAJyWloYHHnjAE1nIDzTsyaTwwWnZDRrGzHDzUyKi4NCqAcC///57q5/QYrFgx44dbQ5Evk1vrPLpS0wAkKGJRGWNDcaqWrmjEBGRF7SqzNx000249NJL8fHHH8Nsbn5g5c6dO/Hoo48iKysLGzdudFvAw4cP48Ybb0RiYiIiIiLQo0cPbNiwwW3PT67RGy0+O5OpQWRYCNrHhnN6NhFRkGjVZaadO3di4cKFeOyxx/CPf/wDXbp0QUpKCsLDw3H8+HHs2rULVVVVuPrqq7Fy5Ur06NHDLeGOHz+OCy64AMOGDcO3334LrVaLvXv3Ij4+3i3PT64zGM24uk8HuWOcVcO2BgN1CXJHISIiD2tVmQkNDcWUKVMwZcoUbNiwAb/88gsOHDiA6upq9OrVCw888ACGDRuGhAT3vnG8+OKLSEtLw9tvv914TKfTufU1yDV6oxk6rW+fmQGATG44SUQUNFweANy/f3/079/fE1ma+O9//4tLL70U1157LX766Sd06NAB9957L+68806vvD45q66141hFjc9uZXAqnSYKfxRyew0iomDQ6hWA5VBYWIiFCxeic+fO+O6773DPPfdgypQpWLp0aYuPsVqtqKiocPog9zCYzIhShUATFSZ3lLPK1HDDSSKiYOHymRlvcjgc6N+/P55//nkAQJ8+fbB9+3YsWrQIEyZMaPYxM2fOxFNPPeXNmEHDYDQjQxPpF9OddRo1DpgssDuET+7uTURE7uPTZ2bat2+P3Nxcp2M5OTk4ePBgi4+ZPn06ysvLGz+Kioo8HTNoFBrN0Gmi5I7RKqnxEXAIgcPHq+WOQkREHubTZ2YuuOAC7N692+nYnj170LFjxxYfo1KpoFKpPB0tKBmMZugSfXO37NOFKBVIT4xEobEK6X6SmYiI2sanz8w88MAD+P333/H8889j3759WLZsGV577TVMmjRJ7mhByWDy3d2ym5OpieK4GSKiINCqMzNz585t9RNOmTKlzWFON2DAAHz++eeYPn06nn76aeh0OsyZMwfjx49322tQ6/nDgnmnytRyEDARUTBoVZl5+eWXnW6XlpbCYrEgLi4OAHDixAlERkYiKSnJrWUGAEaPHo3Ro0e79TnJdZU1dTBWWZHpR2VGp1Hj621H5Y5BREQe1qrLTHq9vvHjueeeQ+/evVFQUICysjKUlZWhoKAAffv2xTPPPOPpvCQTg9GCuMhQxEX6/rTsBjpOzyaShRACBUcrIISQOwoFCZfHzDz++OOYN28eunbt2nisa9euePnll/HYY4+5NRz5Dr3J7BeL5Z0qU6PG4RPVqKmzyx2FKGgIIfDCil247JW1eGOtXu44FCRcLjNHjx6FzWZrctxut6O4uNgtocj36EvNfnWJCQC00SpEqUJgMPHsDJE3OBwCj3+5Hf/bcgSvXN8bL6/ag+938n2BPM/lMjN8+HBMnDgRmzZtajy2ceNG3HPPPcjLy3NrOPId/jaTCQAkSaq/1FTKMkPkaTa7A//8eCt+2WvEx/ecjzG9O+A/Y3vhgeVbsPMIV2Inz3K5zLz11lto164d+vfv37imy8CBA5GcnIw33njDExnJB+iN/ldmgPpxM9xwksizrDY7Ji3bhJ1HK/DR3YPRIS4CADCqZ3vcfVEm7li6HiUVNTKnpEDm8qJ5Wq0W33zzDfbs2YNdu3YBALKzs9GlSxe3hyPfYTCZofOzMTPAyTLDMzNEHlNda8dd725ARY0NH951XpNJApOGdcL+UjPufGcDlk8cjPBQpUxJKZC1edG8jIwMdO3aFZdffjmLTIA7bq7FCUsdMjT+t5Ju/VozVXLHIApIFTV1uPmtP1Brc+D9OwY1O9tRkiS88PceCFUq8OBHW+FwcIYTuZ/LZcZiseD2229HZGQkunXr1rhP0n333YcXXnjB7QFJfnqTGZooFaLDQ+WO4jJOzybyjDJzLca//gfUqhAsuXUgolQtn+hXhSix+KZ+2Hb4BOas2uPFlBQsXC4z06dPx9atW7FmzRqEh4c3Hs/Ly8Py5cvdGo58g8Fohs4Pz8oA9WXmuKUOx821ckchChjFFTUYt3gd0hIi8NpN/RERdvZLR4lRKrw5YQDe/tWALzYf9kJKCiYul5kvvvgC8+fPx5AhQyBJUuPxbt26Yf/+/W4NR75BbzRD54eDfwEgOjwU2mgV9JyeTeQWRWUWXLtoHXqlxWHu9X0QFtL6t5EuydGY948+mP5ZPjYeKPNgSgo2LpeZ0tJSJCUlNTluNpudyg0FDn+dydSAg4CJ3GNfSRWuXbQOF2cnYdbfeyJE6fqwy6Fdk/DIZdm4652NKCqzeCAlBSOX/yX2798fX3/9dePthgLzxhtvYPDgwe5LRj7DX2cyNcjUcBAw0bnacaQc4xavw9/7dcCTV+RCoWj7H68Tzs/AqJ7tcfvS9aisqXNjSgpWLk/Nfv7553HZZZdh586dsNlseOWVV7Bz50789ttv+OmnnzyRkWQkhIDBz3bLPp1Oo8bWQyfkjkHktzYeOI5b3/4T9wzthHuGZrnlOZ8YnYtbl6zHfR9sxhs392/TWR6iBi7/6xkyZAi2bNkCm82GHj16YOXKlUhKSsK6devQr18/T2QkGZVWWVFltfndvkynytRG8TITURv9us+ICW/9iYcu7eq2IgMAIUoFFozvi8PHq/HcNwVue14KTi6fmQGArKwsvP766+7OQj7IYLSgfWx4q2Yr+CqdRg2DyQyHQ5zTqXGiYLNqZzGmfLgZz4zpjr/3S3X788eEh+LNCQNw1au/IlMbhZvO6+j216Dg4PKZmU2bNiE/P7/x9pdffomrrroKjz76KGprOf010BiM/rdb9unSEyJRa3PgGJdTJ2q1/249gvs+2Iz/u7aXR4pMg/TESCy+qR+e/7oAa/eWeux1KLC5XGYmTpyIPXvqFz0qLCzEuHHjEBkZiY8//hgPP/yw2wOSvPR+uMHk6cJCFEhLiOSlJqJW+vDPg5j+6TYsuqkfLuvR3uOvNyAjAc9e1R33vr8J+0oqPf56FHhcLjN79uxB7969AQAff/wxLrroIixbtgxLlizBp59+6u58JDN9qRmZfl5mgIaVgDmjiehs3lhbiOe+LsBbtwzARV20Xnvdv/dLxc2DO+K2JRtQxkUuyUUulxkhBBwOBwBg1apVuPzyywEAaWlpMBqN7k1HsjMEwJkZAMjURHH3bKIzEELglVV7MX/1Prx/5yAMykz0eoYHR3RFt5QY3P3uRlhtdq+/PvmvNq0z8+yzz+Ldd9/FTz/9hFGjRgEA9Ho9kpOT3R6Q5ONwiPo1Zvx0K4NT6bTco4moJUIIPP9NAd774wCW3zUYPVPjZMmhUEiYfV1vVNfZ8ehn2yEEN6Wk1nG5zMyZMwebNm3C5MmTMWPGDHTq1AkA8Mknn+D88893e0CSz7GKGtTaHEhL8P8yk8kNJ4maZXcIPPr5dnyTfwwfTxyMru2iZc0TEabEGxP649d9Riz6qVDWLOQ/XJ6a3bNnT6fZTA3+85//QKn03+m71JTBaEaH+AioQvz/56rTqFFUZoHVZg+I74fIHersDvzz463IP1SOj+8ejJS4CLkjAQCSY8LxxoT+GLd4HXSaSIzs7vlByOTf3LbkYnh4OEJDQ931dOQD9Cb/n5bdoF1MOMJCFNwLhuikmjo77n1/E3Yfq8Tyib5TZBp07xCL2eN648GP6ssW0Zm4XGYUCgWUSmWLHxQ4DH68W/bpFAoJGYnccJIIACy1NtyxdANKK61YftdgaKNVckdq1qXd2uG+4Z1xxzvrcayc60RRy1y+zPT555873a6rq8PmzZuxdOlSPPXUU24LRvLTG824oJNG7hhuk6WN4rgZCnrl1XW4bcl6hCkVeO+OQYhStWkheK+Z+LdM7C+pwh3vrMdHEwcjMsy385I8XP5XMWbMmCbHxo4di27dumH58uW4/fbb3RKM5Kc3mjE+gJYX13EQMAU5U5UVN7/1J5JjwvHq+L4ID/X9s+mSJOG5q3vgxjf/wLTlW/Hq+L7cloSacNuYmfPOOw8//PCDu56OZGZ3CBwss0AXIGNmgPoyw7VmKFgdK6/BdYvXIUOjxqIb+/lFkWkQFqLA4hv7oeBYBV5auVvuOOSD3FJmqqurMXfuXHTo0MEdT0c+4PDxaggBpMb71qDAc6HTcswMBaeDJguuXfwb+nWMx9zr+yAsxG1/x3pNvDoMb04YgPd+P4BPNh6SOw75GJcvM8XHx0OS/jrFJ4RAZWUlIiMj8d5777k1HMlHbzIjPSESIUr/+6XXkkyNGsYqKypq6hATzpl3FBz2lVRi/Bt/4PIe7fH4qFy/vkTTKSkKr47vhzvf2YD0hEgM1CXIHYl8hMtlZs6cOU63FQoFtFotBg0ahPj4eHflIpkZjIGxjcGp4iLDkKAOg8Folm2FUyJv2n64HDe/9SduHJSOB0Z0cfpD1F8N6azBjFE5mPjuBnwx6QJ0DKBL4dR2LpeZCRMmtOp+9957L55++mloNIEzGyaY6I2Bs8bMqRoGAbPMUKDbYCjDrUvWY/KwTph4UZbccdzqxvM6orDUjNuXbsCn95yP2AieaQ12HruG8N5776GiosJTT08epjeaodMGZpnhuBkKdGv3lmLCW3/iXyOzA67INJgxKgfpCZGYvGwTbHaH3HFIZh4rM9wgzL8ZTOaAmsnUgDOaKNCt3HEMd72zEc9e3R03BtDSCqdTKiTMvaEPSiut+Pf/dvA9J8gFzuhOcps6uwOHjlcjIwB2yz5d/YaTVXLHIPKIL7ccxv0fbsHL43rj6j6pcsfxuChVCN6Y0B8rth/D0t8McschGbHMUBNFZRYoFRJSYgNnWnYDnVYNfamZf8VRwFn2x0E8+lk+Ft/UDyO7t5M7jtekxkdi8U398eKK3Vi9u0TuOCQTlhlqon7wb6RfT+FsSUaiGpY6O0orrXJHIXKb137ej5nfFmDJbQPxty5aueN4Xb+O8Xjh7z0wZdlm7D5WKXcckgHLDDURqDOZACA8VImU2AiOm6GAIITA7O/3YOGa/Vh2x3kYkBG8666M6d0Btw3R4fal62Gs4h8rwcZjZebGG29ETEyMp56ePMhgCpzdspuTqeUeTeT/hBB45qsCfPjnQXw0cTB6pMbKHUl2U/M6o096PCa+uxE1dXa545AXtWn70ZqaGmzbtg0lJSVwOJynxF155ZUAgIULF557OpKF3mjG6J4pcsfwmPrp2RwETP7L7hCY8Xk+ftlnxCd3n4/0xMAbrN8WkiThP2N74vrXfscjn27Dy+N6B8RCgXR2LpeZFStW4Oabb4bRaGzyNUmSYLezDfs7g9ES0GdmdBo1ft3X9N8vka8rt9Rh66ETWPbHQewtqcTHdw9G+wAcqH8uwkOVeO3mfrhq/q+Y/+M+3De8s9yRyAtcLjP33Xcfrr32WjzxxBNITk72RCaSUU2dHUfKqwO6zGRqo/Du7wfkjkF0RjV1dhQcrcDWohPYUnQCWw+VQ2+s3zOtf8d4LJ84GJooldwxfVJSdDjevGUArl20DpnaKIzq2V7uSORhLpeZ4uJiTJs2jUUmQB0ssyAiVImk6MD9JZmpUeOgyQKb3RFQG2mS/3I4BAqNZmwtOoGth+rLS8HRCkSpQtArLQ690+Iwpk8H9EqNQ4I6TO64fiGnfQxeub437vtgM1LjI9ArLU7uSORBLpeZsWPHYs2aNcjKCswlsoNdYWn9TKZAvs6cEhcBhSSdXBgwcM9Ake8qqazBloP1xWVrUTm2HjqBWpsD3TvEoldqHG4fokOftHikJUQE9P8XPW14TjKmjeiCO97ZgC8nXYCUOF6SC1Qul5n58+fj2muvxdq1a9GjRw+Ehjpv8DVlyhS3hSPvC/SZTED9MugdEyPrp6AH+PdK8jNbbcg/XF5/qejkx9GKGnROikKv1DiM7N4Oj1yWja7tohHKM4Vud/sQHfaf3JTyk7sHQ61q07wX8nEu/1Q/+OADrFy5EuHh4VizZo3TXw2SJLHM+DmD0RyQ2xicTqdRY39pFYZlJ8kdhQKIze7A7uJKbC0qx5ai49haVI69JZVIig5H77Q49EqLw02DO6JHh1hEh3OnZ2+QJAlPj+mGCW/9ifs/3ILFN/WDMgAXBA12LpeZGTNm4KmnnsIjjzwChYJ/RQQavdGMfh3j5Y7hcTquNUPnSAiBQ8erG8+4bCk6ge1HyhGiUKBnaix6pcVh2iVd0Cs1Du1iw+WOG9RClQosHN8PV7/6K2at2IXpl+fIHYnczOUyU1tbi3HjxrHIBCi90YxMbeBfesnSROGLLYfljkF+5ISl9mRxKT851uUEyqvrkNM+Br3SYjFuQBpeSO+BTE1UQG4F4u9iI0Px5i0DcPWrvyJTq8a4AelyRyI3crnMTJgwAcuXL8ejjz7qiTwkI7PVhpJKa8BuZXAqnpmhM6mps2Pn0YpTBumegMFkQUZiJHqlxWFIJw0mDeuEbikxCA9Vyh2XWkmnUWPh+H64bcl6pCeoMTgrUe5I5CYulxm73Y5Zs2bhu+++Q8+ePZsMAJ49e7bbwpF3GUxmRIeHBMXUT51GjaPlNbDU2hAZxgGBVK/KasMdS9djg+E4YiJC0Ss1Fr3T4nH1yWnR8UHw/41ANzgrEf++Mhd3v7cRX0y6IOAnPAQLl3+L5+fno0+fPgCA7du3O32NUwj9m95oRqYmsKdlN0hUhyE6PAR6oxndUrinDdVbtGY/am0OrP7nUKTGc1p0oBo3IB2FpWbcvmQ9vp16IVQhPLvm71wuM6tXr/ZEDvIBhiCaqixJEjI1apYZanTkRDXe/EWPD+46D2kJgT+jL9g9PDIbX207inX7TRjalbMa/V2bR/Hu27cP3333HaqrqwHUj+wn/6Y3WoJivEyDTG0U9KUcN0P1/vPdblzaLRm9uVJsUFAqJOTlJOH7ncVyRyE3cLnMmEwmDB8+HF26dMHll1+Oo0ePAgBuv/12PPjgg24PSN4TDAvmnUqn4SBgqre16AS+3X4UD43MljsKeVFebjJWFRTzj/EA4HKZeeCBBxAaGoqDBw8iMvKvU7Hjxo3DihUr3BqOvEtvDL4yU8gyE/SEEHj26524Y0gmOnC5+6AySJcIs9WO7Ycr5I5C58jlMTMrV67Ed999h9TUVKfjnTt3xoED3InYX5VX16HMXBs0Y2aAk2WmtApCCA70DGIrth+D3mjB27cOlDsKeVlYiAIXddXi+4Ji9Ejl2Dl/5vKZGbPZ7HRGpkFZWRlUqsDdaTnQGYxmJKjDEBsRPEus6zRqVNTYUGaulTsKycRqs2Pmt7vw4CVdEMU9e4LSJbnJHDcTAFwuMxdeeCHeeeedxtuSJMHhcGDWrFkYNmyYW8OR9+iNZmQkBtcMDrUqBMkxKo6bCWLv/HYAkWFKXNc/Te4oJJOhXZKwt7gSh45b5I5C58DlP0VmzZqF4cOHY8OGDaitrcXDDz+MHTt2oKysDL/++qsnMpIX1I+XiZI7htdlaqJQaDSjf0aC3FHIy8rMtZj74168Or4vNx4MYrGRoRiQkYAfCkow4fwMueNQG7l8ZqZ79+7Ys2cPhgwZgjFjxsBsNuOaa67B5s2bkZWV5YmMjV544QVIkoSpU6d69HWCUf1MpuA6MwNwW4NgNveHvejfMR4XdtbKHYVk1jCrifxXmy4Sx8bGYsaMGe7Ockbr16/H4sWL0bNnT6++brAwGM0YkZssdwyvy9SoscFwXO4Y5GX7SqrwwZ8H8dV9Q+SOQj5gRE4yXvi2ABU1dYgJD55xg4GkTWWmpqYG27ZtQ0lJCRwOh9PXrrzySrcEO1VVVRXGjx+P119/Hc8++6zbnz/YCSFQGGTTshvoNGp8tKFI7hjkZS98W4Dr+qehc3K03FHIB6QnRiJTE4Wfdpfiil4pcsehNnC5zKxYsQI333wzjEZjk69JkgS73e6WYKeaNGkSRo0ahby8vLOWGavVCqvV2ni7ooLrB5xNmbkWlTW2oFr9t4FOo4bBZIHdIThuIkj8us+IPwrLsOahoXJHIR+Sl5uEVQXFLDN+yuUxM/fddx+uvfZaHD16FA6Hw+nDE0Xmww8/xKZNmzBz5sxW3X/mzJmIjY1t/EhL4yyFszGYzEiKVkEdhFNT0xIi4XAIHDlRLXcU8gK7Q+DZrwsw+eJOSIziUhL0l7ycZKzeVYI6u+Psdyaf43KZKS4uxrRp05Cc7PnxFUVFRbj//vvx/vvvIzw8vFWPmT59OsrLyxs/iop4CeFs9EZLUC2Wd6pQpQLpCZEcBBwkPt14CJU1dZy1Qk30So2DKlSJ9foyuaNQG7hcZsaOHYs1a9Z4IEpTGzduRElJCfr27YuQkBCEhITgp59+wty5cxESEtLsmSCVSoWYmBinDzozvbEKmUFaZgDu0RQszFYbXlq5G49clo3wUKXcccjHKBo2nuSsJr/k8nWF+fPn49prr8XatWvRo0cPhIY6j/yeMmWK28INHz4c+fn5TsduvfVWZGdn41//+heUSv5CcgeD0RLUS3mzzASHxT8XIjU+AqN6tJc7CvmovJxk/Pt/O/DE6FxuceJnXC4zH3zwAVauXInw8HCsWbPG6QcuSZJby0x0dDS6d+/udEytViMxMbHJcWo7vdEc1IPedFo1Vmw/JncM8qCj5dV4/edCvH/nIL5JUYsu6KSBsbIWu4srkd2OZ/X9ictlZsaMGXjqqafwyCOPQKFw+SoV+RghxMkF83iZiQLXf77bjRG5yeibHi93FPJh4aFKXNhZg1U7i1lm/IzLZaa2thbjxo2Trch4a7xOsCiptKK6zo6OQbYv06mytFE4fKIaNXV2jqUIQNsOncA3+UexatpFckchP5CXm4z3/ziIyRd3ljsKucDlRjJhwgQsX77cE1lIBnqjGSmxEUH9Jp4UrUJEqBIHy7jRXKARon4q9m0X6JAaH7yFnVrv4uwkbD9cjpKKGrmjkAtcPjNjt9sxa9YsfPfdd+jZs2eTAcCzZ892WzjyPIPRjIwg3JPpVJIkQadRo7DUjC5cETagrNxZjMLSKrw5ob/cUchPaKJU6J0Wh1UFJfjHoHS541AruVxm8vPz0adPHwDA9u3bnb7GgXX+Rx+k2xicTqdRo9BYJXcMcqNamwMzvynAAyO6IJr77ZALRpzceJJlxn+4XGZWr17tiRwkE73RjIG6BLljyC5To4a+lIOAA8m7vx9AWIgC4/pzFXByTV5OMmZ/vweWWhsiw4JvZXR/xOlIQS7YZzI1yNRGcUZTADlhqcXcH/ZixqhchCj5a45ck6VVo0NcBNbubboHIfkm/r88iDkcAgZT8G5lcCpOzw4sr/ywF73T4nBRF63cUcgPSdLJ1YB3cjVgf8EyE8SOlFfD7hBI4ywPZGjUMJlrUW6pkzsKnaPC0ios++MgZozKkTsK+bERue3w464S2B1C7ijUCiwzQcxgtCA1PgJhIfxnEBsRCk1UGPQmnp3xdzO/3YWx/VI5M43OSd/0OAghsPngcbmjUCvwXSyI6U1mZCTyElOD+unZnNHkz9btN+H3/SY8MKKL3FHIz4UoFRiWzY0n/QXLTBDTl3Lw76k4bsa/ORwCz369E/cMy4ImSiV3HAoAl+Qmc9yMn2CZCWKcyeQsUxuFQpYZv/XZ5sM4YanDbRfo5I5CAeLCzlocKqvGfp6x9XksM0GsfvVflpkGOq4147cstTb857td+Ndl2UG9NQe5l1oVgvM7JeIHXmryeSwzQcpmd+BgmQU6jplplHnyMpMQnL3gb177uRApcRG4omd7uaNQgMnLScaqnSVyx6CzYJkJUoeOV0OSgA7xEXJH8RnpiZGosdlRXGGVOwq54Fh5DV77uRCPjcrllirkdnk5ydh48DjKzLVyR6EzYJkJUnqTGekJkVAq+Mu/gSpEidT4CM5o8jMvrdyNi7OT0K9jvNxRKAC1iw1Ht5QY/LiLZ2d8GctMkDJwg8lmZWo4CNifbD9cjq+2HcG/RmbLHYUCWP2lJo6b8WUsM0FKb+QaM83h9Gz/IYTAc18X4JbzdUhL4CrW5Dl5Ocn4eW8paursckehFrDMBCm90QydlmXmdJlalhl/saqgBLuLK3HvsCy5o1CAy2kfjfjIMKzbb5I7CrWAZSZIGUxmzmRqBs/M+IdamwPPf1OAB0Z0QUx4qNxxKMBJkoQRuclcDdiHscwEIavNjsPHq7nGTDN0GjUOlllQa3PIHYXO4P0/DkCpkHDDgDS5o1CQyMtJxg8FxXBw40mfxDIThIrKLAgLUaBdTLjcUXxOSmwEQhQSio5b5I5CLThhqcWcVXsxY1QOQpT8FUbeMVCXAIvVju1HyuWOQs3gb4IgpDdakJGohoLTsptQKCSuBOzj5v24Dz1TYzG0i1buKBREwkIUGJqdxL2afBTLTBAycCbTGXHcjO8yGM147/cDmDEqhwvkkdfl5bDM+CqWmSCkN3FPpjPRadRca8ZHvfDtLlzTNxXZ7WLkjkJBaGiXJOwrqUJRGS9D+xqWmSCkLzUjk2WmRfVnZrgKsK/5o9CEtXtLMW1EF7mjUJCKjQzFQF0CN570QSwzQcjAMzNnlKlVo5BjZnyKwyHw7NcFuHdYJ2ijVXLHoSDGKdq+iWUmyFTX2nG0vAYZGq6Y2pJMTRRKKq2ostrkjkInfbHlMExVVtw+RCd3FApyeTnJ+KOwDOXVdXJHoVOwzAQZg8kMdZgS2ij+dduSeHUY4iJDYeC4GZ9QXWvHrBW78a/LshEeqpQ7DgW5tIRIdEqKwk97SuWOQqdgmQkyhpPbGHAmyJlxELDveH1tIZJjw3FFzxS5oxAB4MaTvohlJsjoTZyW3Rpca8Y3lFTUYPFP+/H4qByui0Q+Y0RuMlbvLkGdnSuF+wqWmSBjMJqh4+Dfs8rUqFHIGU2y+7+VezC0axL6ZyTIHYWoUY8OsYgIVeJPfZncUegklpkgo2eZaRWdJooL58ls55EKfLn1MP41MlvuKEROFAoJw3OSuYCeD2GZCTJ6o4XTslshU1t/mUkIbionByEEnv16J24enIH0RM68I98zIjcJqwqK+TvCR7DMBJHKmjoYq6zQcczMWWUkqlFptcFYVSt3lKD0464SFBytwKRhneSOQtSs87M0MFXVYtexSrmjEFhmgorBaEFsRCji1WFyR/F5EWFKpMSG81KTDOrsDjz3TQEeGNEFsRGhcschalZ4qBJ/66LhrCYfwTITRPQmjpdxhU7LbQ3ksOyPg5AA3DAwXe4oRGeUl5OMVVwN2CewzAQRzmRyjU7DbQ28rby6DnNW7cGMUTkIVfLXE/m2i7OTsP1IBYorauSOEvT42yKIGIxcY8YVmZooLpznZQtW70O3lFgM65okdxSis0qMUqFvehzPzvgAlpkgUmg0c08mF9RfZmKZ8ZYDJjOW/mbAjFE5XKGa/AZXA/YNLDNBxGAyI1MTJXcMv5GpUeOAyQy7g1MvveHFFbtwVe8OyGkfI3cUolbLy03Gr/tNMHNjWlmxzASJE5ZanLDU8cyMCzrERQAADh+vljlJ4FtvKMNPu0vx4CVd5I5C5JIsbRRS4yOwdq9R7ihBjWUmSOiNZmiiwhAdzqmurRWiVCA9IRL7OaPJoxwOgWe/2om7L8pCUky43HGIXDaCqwHLjmUmSHAbg7bRaaK44aSH/W/bEZRUWnHHhZlyRyFqk7zcZPy4q5iXpGXEMhMkOJOpbbI4CNijaursePHbXXh4ZFdEhCnljkPUJn3T4yFJEjYdPC53lKDFMhMk9CbuydQWOg3LjCe9+YsemmgVxvTqIHcUojZTKiRcnJ3EWU0yYpkJEnpjFS8ztQHLjOeUVNbg1dX78NioXCgUnIpN/i2P42ZkxTITBIQQMBgtLDNtoNOqcfhENWrq7HJHCTgvf78Hf+uixUBdgtxRiM7ZhZ01OHSiGvtLOWFADiwzQcBYVYsqq41jZtpAG6VClCqEZ2fcrOBoBT7bdBiPXJYtdxQit1CrQnBBViIvNcmEZSYIGExmtIsJ5wDLNpAkCZkcBOxWQgg8/00BbjqvIzqyYFMAGZHbjlsbyIRlJgjoS7mNwbnguBn3WrOnFPmHy3HfxZ3ljkLkVsNzkrDp4AmYqqxyRwk6LDNBQG8yQ8dtDNqMu2e7j83uwHNfF2Dq8M6IjeQCjhRYkmPC0T0lBj/uKpE7StAJkTsAeZ7BaEaf9Di5Y/gtnUaNn/eUyh2jCSEE6uwCNTY7aursqLU5AAAKSYJCkiBJgCSdcvvk1yRFwzFAgnTKfeovqzV89oQP1hfB4RAYf15Hjzw/kdzycpKxqqAY1/ZPkztKUGGZCQJ6oxlX9+E6Hm2VqYlC4VkuMwkhYLU5YK1zoMZmb/xcU2dHTZ0DVlv95/rbdlhtjmY/N9zf6ZjNAWsz97Ha7Dh1wVFJAiQA7lqE9PRyozhZehpLkQQoFKfebrivc6E69fPR8mrMu6EvQpU8KUyBaUS3ZLy6Zj9q6uwID+U4RW9hmQlwDoeo3y1by4GWbaXTqnHCUodrF/3mVFhOLx6nClFICA9VQhWiqP8cqkB4iBLhoQqoTn4OD1U63SciTIn4yDCnY3/d76/H1X9uep9TC4IQAg7x12eHEBANn3Hys8P5ttN9Wvh86nMKCDgcLTx3432dnztKFYJeaXHe/PEReVXX5GgkRoXht/1GXJydLHecoMEyE+CKK2tQa3MgLYEDgNsqShWC12/ujypr3clCUl9OWiol4SEKhMh85kGSJCgloP5cDRF5iyRJJxfQK2GZ8SKWmQCnN5qREhcBVQhPd56LEbn8pURErTMiNxkPLN8Ch6M7V7f2El64DnDcLZuIyLsG6hJQXWdH/uFyuaMEDZaZAGdgmSEi8qpQpQLDuiZxryYv8vkyM3PmTAwYMADR0dFISkrCVVddhd27d8sdy2/ojRZuY0BE5GV5uclcDdiLfL7M/PTTT5g0aRJ+//13fP/996irq8Mll1wCs5mLmLWGwcQzM0RE3nZRFy32l1ahqMwid5Sg4PMDgFesWOF0e8mSJUhKSsLGjRvxt7/9TaZU/mFPcSUOmizolMTVf4mIvCk2IhSDdIlYVVCMWy/QyR0n4Pn8mZnTlZfXD6hKSEho9utWqxUVFRVOH8GovLoOd72zAXdflMlp2UREMsjL4bgZb/GrMuNwODB16lRccMEF6N69e7P3mTlzJmJjYxs/0tKCb0lph0Ng6oebodOoMTWvi9xxiIiC0vCcZPypL0O5pU7uKAHPr8rMpEmTsH37dnz44Yct3mf69OkoLy9v/CgqKvJiQt8w54e90BvNmHN9H65xQEQkk7SESHRKisKaPdx40tN8fsxMg8mTJ+Orr77Czz//jNTU1Bbvp1KpoFKpvJjMt3y/sxhvri3EZ/degNgI7kpMRCSnS3KTsaqgBGN6c388T/L5MzNCCEyePBmff/45fvzxR+h0HEjVkv2lVZi2fAtmje2Fru2i5Y5DRBT08nKTsWZXSeOu9uQZPl9mJk2ahPfeew/Lli1DdHQ0jh07hmPHjqG6ulruaD6lsqZ+wO8/zkvHqJ7t5Y5DREQAuqfEIlKlxJ/6MrmjBDSfLzMLFy5EeXk5hg4divbt2zd+LF++XO5oPsPhEPjnx1vRPjYCD13SVe44RER0kkIhYXgOF9DzNJ8vM0KIZj9uueUWuaP5jIU/7ceOIxWYd0Mf2XdrJiIiZyNyk/H9zmIIIeSOErD4zufnVu8uwYLV+7Doxn6IV4fJHYeIiE4zODMRxy21KDhaKXeUgMUy48cOmMy4/4PNeO7q7ujeIVbuOERE1IzwUCX+1lnLS00exDLjpyy1Nkx8dyP+3i8VV/dpeao6ERHJbwQ3nvQolhk/JITAw59sQ2xEKB69PEfuOEREdBbDspOw40gFjpXXyB0lILHM+KHX1xZi44HjWDC+L0I54JeIyOclqMPQLz2eZ2c8hO+EfubXfUa8/P1eLLqxHzRRwbvSMRGRv8nLTWKZ8RCWGT9y6LgFk5dtwlNXdkOvtDi54xARkQtG5LbDb/tMMFttckcJOCwzfqKmzo6J727EqJ7tcd2A4NsJnIjI3+k0aqQlRODnPaVyRwk4LDN+QAiBRz/PR3ioEk+M7iZ3HCIiaqO83GR8z0tNbscy4weW/mbA2r1GvDq+L8JC+CMjIvJXI3KSsXpXCWx2bjzpTnxn9HF/FJrw4ordWDi+L5JjwuWOQ0RE56BPejwUkoRNB0/IHSWgsMz4sKPl1Zi0bBMeHZWD/hkJcschIqJzpFRIuDg7Cd/vPCZ3lIDCMuOjrDY77n5vEy7OTsKNg9LljkNERG6Sx40n3Y5lxkc9+eUOQAg8PaY7JEmSOw4REbnJhZ01OFpeg/2lZrmjBAyWGR+07I+D+H5nMRbe2A/hoUq54xARkRtFhoVgSCcNF9BzI5YZH7PxwHE889VOLBjfFylxEXLHISIiD2i41ETuwTLjQ0oqa3Dv+xvx0KVdcV5motxxiIjIQ4ZnJ2FL0QkYq6xyRwkILDM+otbmwKT3N2FwZiJuvSBD7jhERORBSTHh6N4hFj/uKpE7SkBgmfERz369E2arHTOv6ckBv0REQeCS3GSs4qUmt2CZ8QEfbyjCf7ceweKb+iEijAN+iYiCQV5OMtbuNaKmzi53FL/HMiOzbYdO4Ikvd2DeDX2QlhApdxwiIvKSLslR0ESH4dd9Rrmj+D2WGRmZqqy4+92NuD+vMy7srJU7DhEReZEkScjLSeYUbTdgmZGJze7ApGWb0Cc9HhP/lil3HCIiksGI3GSsKiiBw8HVgM8Fy4xMXvh2F8rMtZg1lgN+iYiC1YCMBFjr7Nh66ITcUfway4wMvtxyGB9tKMJrN/WHWhUidxwiIpJJqFKBYdlJvNR0jlhmvGznkQpM/ywfc67vjQyNWu44REQks7ycZKzayfVmzgXLjBedsNRi4nsbcPdFWbg4O1nuOERE5AMu6qpFobEKB00WuaP4LZYZL7E7BO77YDOy28Vg8rBOcschIiIfERMeivMyE/E9LzW1GcuMl/zfyt04fKIas6/rBYWCA36JiOgv9ZeaWGbaimXGC77NP4p31x3Aazf1R3R4qNxxiIjIxwzPScJ6QxnKLXVyR/FLLDMetqe4Eg99sg0vXdcLnZKi5I5DREQ+KDU+El2So7FmDwcCtwXLjAeVV9dh4rsbccv5Gbi0Wzu54xARkQ/Ly03GSl5qahOWGQ9xOASmLd+CjomReGBEF7njEBGRjxuRk4yfdpei1uaQO4rfYZnxkFd+2It9pVV4ZVwfKDngl4iIzqJ7hxhEqULwh94kdxS/wzLjAd/vLMYbawux+KZ+iI3kgF8iIjo7SZKQl5vEWU1twDLjZvtLqzBt+Ra8OLYnstvFyB2HiIj8SF5OMr7fWQwhuPGkK1hm3KjKasPEdzfiH4PSMbpnitxxiIjIzwzOSkR5dR12Hq2QO4pfYZlxEyEE/vnRViTHqPDQpV3ljkNERH5IFaLERV213KvJRSwzbvLqmv3IP1yOeTf0RYiS/1mJiKht8nKSuYu2i/iu6wZrdpdgwep9WHxTPySow+SOQ0REfmxY1yTsPFqBo+XVckfxGywz5+igyYL7P9yCZ6/qju4dYuWOQ0REfi5eHYZ+HeOxqoCXmlqLZeYcWGptuOvdDbi6Twdc0zdV7jhERBQgLsnlxpOuYJlpIyEE/vVpPmIiQjFjVI7ccYiIKIAMz0nGuv0mVFltckfxCywzbVRT50CoUsKCf/RFKAf8EhGRG+k0aqQnRuLnPaVyR/ELfBduo4gwJWZf1xvaaJXcUYiIKADl5fBSU2uFyB2AiIiImhqRm4xb3v4TT3653a3P64m1hR+9PAfhoUoPPHPrsMwQERH5oD5pcbjrwkzPjJsJsP2PWWaIiIh8kEIh4b7hneWO4Rc4ZoaIiIj8GssMERER+TWWGSIiIvJrLDNERETk11hmiIiIyK+xzBAREZFfY5khIiIiv8YyQ0RERH6NZYaIiIj8ml+UmQULFiAjIwPh4eEYNGgQ/vzzT7kjERERkY/w+TKzfPlyTJs2DU8++SQ2bdqEXr164dJLL0VJSYnc0YiIiMgH+HyZmT17Nu68807ceuutyM3NxaJFixAZGYm33npL7mhERETkA3y6zNTW1mLjxo3Iy8trPKZQKJCXl4d169bJmIyIiIh8hU/vmm00GmG325GcnOx0PDk5Gbt27Wr2MVarFVartfF2RUWFRzMSERGRvHy6zLTFzJkz8dRTTzU5zlJDRETkPxret4UQZ72vT5cZjUYDpVKJ4uJip+PFxcVo165ds4+ZPn06pk2b1nj78OHDyM3NRVpamkezEhERkftVVlYiNjb2jPfx6TITFhaGfv364YcffsBVV10FAHA4HPjhhx8wefLkZh+jUqmgUqkab0dFRaGoqAjR0dGQJMkbsYmIiOgcCSFQWVmJlJSUs97Xp8sMAEybNg0TJkxA//79MXDgQMyZMwdmsxm33nprqx6vUCiQmprq4ZRERETkbmc7I9PA58vMuHHjUFpaiieeeALHjh1D7969sWLFiiaDgomIiCg4SaI1I2uIiIiIfJRPrzNDREREdDYsM0REROTXWGaIiIjIr7HMEBERkV9jmSEiIiK/xjJDREREfo1lhoiIiPwaywwRERH5NZYZIiIi8mssM0REROTXWGaIiIjIr7HMEBERkV/7f+MGkhj4mfydAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mw.save_metric_plot(file_path=f\"/tmp/{instance_id}_plot.png\", query_kwargs=query_kwargs)\n",
    "mw.save_metric_csv(file_path=f\"/tmp/{instance_id}_metric.csv\", query_kwargs=query_kwargs)\n",
    "mw.save_metric_json(file_path=f\"/tmp/{instance_id}_metric.json\", query_kwargs=query_kwargs)\n",
    "mw.save_response_json(file_path=f\"/tmp/{instance_id}_response.json\", query_kwargs=query_kwargs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual EC2 querying\n",
    "\n",
    "For users that require more control over the EC2 instance query settings, the `query_ec2_metrics` method can be used to manually query the EC2 instance. For instance it allows to fine tune the query period settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RequestId': '256d262e-488a-42bc-9197-2fcd4df82e98',\n",
       " 'HTTPStatusCode': 200,\n",
       " 'HTTPHeaders': {'x-amzn-requestid': '256d262e-488a-42bc-9197-2fcd4df82e98',\n",
       "  'content-type': 'text/xml',\n",
       "  'content-length': '1596',\n",
       "  'date': 'Wed, 19 Jul 2023 18:35:04 GMT'},\n",
       " 'RetryAttempts': 0}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FINE_TUNED_SETTINGS = {\n",
    "    \"days\": 5,\n",
    "    \"hours\": 0,\n",
    "    \"minutes\": 0,\n",
    "    \"stat\": \"Maximum\",\n",
    "    \"period\": 60,\n",
    "}\n",
    "response = mw.query_ec2_metrics(**FINE_TUNED_SETTINGS)\n",
    "\n",
    "response[\"ResponseMetadata\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `TimedMetric` dataclass\n",
    "\n",
    "Internally, the package uses `TimedMetric` dataclass to store the metric data. This dataclass is used to store the metric data and provide a convenient interface to access the data. It can be also used to interact with the metric data by the user."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'cloudwatcher.metric_handlers.TimedMetric'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[485814272.0,\n",
       " 4685066240.0,\n",
       " 6207594496.0,\n",
       " 4992217088.0,\n",
       " 4720185344.0,\n",
       " 2435854336.0,\n",
       " 2444738560.0,\n",
       " 2400739328.0,\n",
       " 11007488000.0]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response = mw.query_ec2_metrics(**FINE_TUNED_SETTINGS)\n",
    "timed_metric = mw.timed_metric_factory(response)[0]\n",
    "print(timed_metric.__class__)\n",
    "timed_metric.values[1:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `LogWatcher`: convenient interface to AWS CloudWatch logs\n",
    "\n",
    "`LogWatcher` can be used to interact with AWS CloudWatch logs.\n",
    "\n",
    "### `LogWatcher` initialization\n",
    "\n",
    "As described in the Login credentials section, the AWS credentials can be sourced from environment variables: `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`.\n",
    "Alternatively, you can pass the values as arguments to the `LogWatcher` constructor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cloudwatcher.logwatcher import LogWatcher\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "lw = LogWatcher(\n",
    "    log_group_name=os.environ.get(\"LOG_GROUP_NAME\"),\n",
    "    log_stream_name=os.environ.get(\"LOG_STREAM_NAME\"),\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importantly, you can also provide the start token for the log, which will be used to determine the starting point of the log query."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Log streaming\n",
    "\n",
    "`LogWatcher` provides a convenient interface to stream the logs from AWS CloudWatch. There are 2 relevant parameters in `stream_cloudwatch_logs` method:\n",
    "\n",
    "- `events_limit` - the maximum number of events to be returned. If the value is set to `None`, 1000 events are returned\n",
    "- `max_retry_attempts` - the maximum number of retry attempts to be made if the query results with an empty log\n",
    "\n",
    "The `stream_cloudwatch_logs` method returns a generator that yields the log events, for example in a `for` loop. In the example below , we use `next` to get the first event from the generator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LogEventsList(events=[LogEvent(message='[2023-07-19 12:34:48,735 - INFO] Nephele, developed by BCBB/OCICB/NIAID/NIH version: 2.27.1, tag: Nephele_2023_July_19, commit: dce18e5', timestamp=datetime.datetime(2023, 7, 19, 12, 34, 48, 833000)), LogEvent(message='[2023-07-19 12:34:48,736 - INFO] Python version: 3.9.2', timestamp=datetime.datetime(2023, 7, 19, 12, 34, 48, 833000))], next_forward_token='f/37683453325894048129959718411585392236426944928309968897/s', next_backward_token='b/37683453325894048129959718411585392236426944928309968896/s')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "streamer = lw.stream_cloudwatch_logs(events_limit=2, max_retry_attempts=2)\n",
    "next(streamer)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The log events are returned as a custom `LogEventsList` object, which conists of a list of `LogEvents` and tokens. The next token (`LogEventsList.next_forward_token`) can be used to get the next batch of log events. The token can be provided to the `LogWatcher` constructor to start streaming from the last event.\n",
    "\n",
    "### Retrieving all logs\n",
    "\n",
    "Alternatively, the `return_formatted_logs` method can be used to retrieve all the logs. This method returns a `Tuple[str,str]`, where the first element is the formatted log and the second element is the next token. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[19-07-2023 12:34:48 UTC] Nephele, developed by BCBB/OCICB/NIAID/NIH version: 2.27.1, tag: Nephele_2023_July_19, commit: dce18e5\n",
      "[19-07-2023 12:34:48 UTC] Python version: 3.9.2\n",
      "[19-07-2023 12:34:48 UTC] Current time: 2023-07-19 12:34\n",
      "[19-07-2023 12:34:49 UTC] Pipeline name: DADA2\n",
      "[19-07-2023 12:34:49 UTC] Job Description:\n",
      "[19-07-2023 12:34:49 UTC] Job parameters\n",
      "[19-07-2023 12:34:49 UTC] job_id: b73de8bfdd22\n",
      "[19-07-2023 12:34:49 UTC] inputs_dir: None\n",
      "[19-07-2023 12:34:49 UTC] outputs_dir: None\n",
      "[19-07-2023 12:34:49 UTC] map_file: <_io.TextIOWrapper name='/nephele_data/inputs/N2_16S_example_mapping_file_3_corrected.txt' mode='r' encoding='UTF-8'>\n",
      "[19-07-2023 12:34:49 UTC] data_type: PE\n",
      "[19-07-2023 12:34:49 UTC] wurlitzer_stdout: file\n",
      "[19-07-2023 12:34:49 UTC] wurlitzer_stderr: file\n",
      "[19-07-2023 12:34:49 UTC] ion_torrent: False\n",
      "[19-07-2023 12:34:49 UTC] trimleft_fwd: 0\n",
      "[19-07-2023 12:34:49 UTC] trimleft_rev: 0\n",
      "[19-07-2023 12:34:49 UTC] maxee: 5\n",
      "[19-07-2023 12:34:49 UTC] trunclen_fwd: 0\n",
      "[19-07-2023 12:34:49 UTC] trunclen_rev: 0\n",
      "[19-07-2023 12:34:49 UTC] truncq: 4\n",
      "[19-07-2023 12:34:49 UTC] just_concatenate: False\n",
      "[19-07-2023 12:34:49 UTC] maxmismatch: 0\n",
      "[19-07-2023 12:34:49 UTC] trim_overhang: False\n",
      "[19-07-2023 12:34:49 UTC] chimera: True\n",
      "[19-07-2023 12:34:49 UTC] ref_db: sv138.1\n",
      "[19-07-2023 12:34:49 UTC] taxmethod: rdp\n",
      "[19-07-2023 12:34:49 UTC] sampling_depth: None\n",
      "[19-07-2023 12:34:49 UTC] pseudopool: False\n",
      "[19-07-2023 12:34:49 UTC] minboot: 80\n",
      "[19-07-2023 12:34:49 UTC] allowmultiplespecies: False\n",
      "[19-07-2023 12:34:49 UTC] Results manager initialized. Results registry path: /mnt/EFS/user_uploads/b73de8bfdd22/outputs/b73de8bfdd22_results_registry.json\n",
      "[19-07-2023 12:34:49 UTC] Checking Mapfile for Gzipped inputs.\n",
      "[19-07-2023 12:34:49 UTC] Gzipped files listed in map file, attempting to rm .gz extension.\n",
      "[19-07-2023 12:34:51 UTC] Done. Attempting file decompression.\n",
      "[19-07-2023 12:34:51 UTC] Finished decompression.\n",
      "[19-07-2023 12:34:55 UTC] Skipping FASTQ file validation\n",
      "[19-07-2023 12:35:04 UTC] Reference DB (sv138.1) checksum: 6b41db7139834c71171f8ce5b5918fc6\n",
      "[19-07-2023 12:35:05 UTC] Taxonomy assignemnt DB checksum: f21c2d97c79ff07c17949a9622371a4c\n",
      "[19-07-2023 12:35:05 UTC] Running dada2nephele.R with command:\n",
      "[19-07-2023 12:35:09 UTC] Rscript /usr/local/src/nephele2/pipelines/DADA2/dada2nephele/R/dada2nephele.R  --datadir /nephele_data/inputs/ --outdir /nephele_data/outputs/ --mapfile /nephele_data/outputs/N2_16S_example_mapping_file_3_corrected.txt.no_gz --logfilename /var/log/job.log --nthread 12 --maxEE 5 --truncQ 4 --maxMismatch 0 --chimera --data_type PE --minBoot 80 --no_MultipleSpecies --trimLeft_R1 0 --trimLeft_R2 0 --truncLen_R1 0 --truncLen_R2 0 --taxmethod rdp --refdb /mnt/EFS/dbs/dada2_silva_v138.1/silva_nr99_v138.1_train_set.fa.gz --refdb_species /mnt/EFS/dbs/dada2_silva_v138.1/silva_species_assignment_v138.1.fa.gz\n",
      "[19-07-2023 12:35:16 UTC] R version 4.3.1 (2023-06-16)\n",
      "[19-07-2023 12:35:16 UTC] Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "[19-07-2023 12:35:16 UTC] Running under: Debian GNU/Linux 11 (bullseye)\n",
      "[19-07-2023 12:35:16 UTC] Matrix products: default\n",
      "[19-07-2023 12:35:16 UTC] BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0\n",
      "[19-07-2023 12:35:16 UTC] LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0\n",
      "[19-07-2023 12:35:16 UTC] locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C\n",
      "[19-07-2023 12:35:16 UTC] [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C\n",
      "[19-07-2023 12:35:16 UTC] time zone: America/New_York\n",
      "[19-07-2023 12:35:16 UTC] tzcode source: system (glibc)\n",
      "[19-07-2023 12:35:16 UTC] attached base packages:\n",
      "[19-07-2023 12:35:16 UTC] [1] stats     graphics  grDevices utils     datasets  methods   base\n",
      "[19-07-2023 12:35:16 UTC] other attached packages:\n",
      "[19-07-2023 12:35:16 UTC] [1] dada2_1.28.0 Rcpp_1.0.11  docopt_0.7.1\n",
      "[19-07-2023 12:35:16 UTC] loaded via a namespace (and not attached):\n",
      " [1] utf8_1.2.3                  generics_0.1.3             \n",
      " [3] bitops_1.0-7                stringi_1.7.12             \n",
      " [5] jpeg_0.1-10                 lattice_0.20-45            \n",
      " [7] magrittr_2.0.3              grid_4.3.1                 \n",
      " [9] RColorBrewer_1.1-3          iterators_1.0.14\n",
      "[19-07-2023 12:35:16 UTC] [11] foreach_1.5.2               plyr_1.8.8\n",
      "[19-07-2023 12:35:16 UTC] [13] Matrix_1.5-3                GenomeInfoDb_1.36.1\n",
      "[19-07-2023 12:35:16 UTC] [15] fansi_1.0.4                 scales_1.2.1\n",
      "[19-07-2023 12:35:16 UTC] [17] Biostrings_2.68.1           codetools_0.2-19\n",
      "[19-07-2023 12:35:16 UTC] [19] cli_3.6.1                   ShortRead_1.58.0\n",
      "[19-07-2023 12:35:16 UTC] [21] rlang_1.1.1                 crayon_1.5.2\n",
      "[19-07-2023 12:35:16 UTC] [23] XVector_0.40.0              Biobase_2.60.0\n",
      "[19-07-2023 12:35:16 UTC] [25] munsell_0.5.0               DelayedArray_0.26.6\n",
      "[19-07-2023 12:35:16 UTC] [27] S4Arrays_1.0.4              tools_4.3.1\n",
      "[19-07-2023 12:35:16 UTC] [29] parallel_4.3.1              reshape2_1.4.4\n",
      "[19-07-2023 12:35:16 UTC] [31] deldir_1.0-9                BiocParallel_1.34.2\n",
      "[19-07-2023 12:35:16 UTC] [33] dplyr_1.1.2                 interp_1.1-4\n",
      "[19-07-2023 12:35:16 UTC] [35] colorspace_2.1-0            ggplot2_3.4.2\n",
      "[19-07-2023 12:35:16 UTC] [37] GenomeInfoDbData_1.2.10     Rsamtools_2.16.0\n",
      "[19-07-2023 12:35:16 UTC] [39] hwriter_1.3.2.1             SummarizedExperiment_1.30.2\n",
      "[19-07-2023 12:35:16 UTC] [41] BiocGenerics_0.46.0         png_0.1-8\n",
      "[19-07-2023 12:35:16 UTC] [43] vctrs_0.6.3                 R6_2.5.1\n",
      "[19-07-2023 12:35:16 UTC] [45] matrixStats_1.0.0           stats4_4.3.1\n",
      "[19-07-2023 12:35:16 UTC] [47] lifecycle_1.0.3             stringr_1.5.0\n",
      "[19-07-2023 12:35:16 UTC] [49] zlibbioc_1.46.0             S4Vectors_0.38.1\n",
      "[19-07-2023 12:35:16 UTC] [51] IRanges_2.34.1              pkgconfig_2.0.3\n",
      "[19-07-2023 12:35:16 UTC] [53] RcppParallel_5.1.7          pillar_1.9.0\n",
      "[19-07-2023 12:35:16 UTC] [55] gtable_0.3.3                glue_1.6.2\n",
      "[19-07-2023 12:35:16 UTC] [57] tibble_3.2.1                GenomicAlignments_1.36.0\n",
      "[19-07-2023 12:35:16 UTC] [59] GenomicRanges_1.52.0        tidyselect_1.2.0\n",
      "[19-07-2023 12:35:16 UTC] [61] MatrixGenerics_1.12.2       latticeExtra_0.6-30\n",
      "[19-07-2023 12:35:16 UTC] [63] compiler_4.3.1              import_1.3.0\n",
      "[19-07-2023 12:35:16 UTC] [65] RCurl_1.98-1.12\n",
      "[19-07-2023 12:35:16 UTC] Taxonomic Reference Database\n",
      "[19-07-2023 12:35:16 UTC] /mnt/EFS/dbs/dada2_silva_v138.1/silva_nr99_v138.1_train_set.fa.gz\n",
      "[19-07-2023 12:35:16 UTC] /mnt/EFS/dbs/dada2_silva_v138.1/silva_species_assignment_v138.1.fa.gz\n",
      "[19-07-2023 12:35:16 UTC] Reading in map file  /nephele_data/outputs/N2_16S_example_mapping_file_3_corrected.txt.no_gz\n",
      "[19-07-2023 12:35:16 UTC] Printing dada algorithm options.\n",
      " [1]           16        FALSE           -8         TRUE         TRUE\n",
      " [6]                      0.42            5            0           10\n",
      "[19-07-2023 12:35:16 UTC] [11]            1            1            1           -4 0.000000....\n",
      "[19-07-2023 12:35:16 UTC] [16] 0.000000....       0.0001          Inf            2            2\n",
      "[19-07-2023 12:35:16 UTC] [21]         TRUE         TRUE         TRUE\n",
      "[19-07-2023 12:35:16 UTC] Paired End\n",
      "[19-07-2023 12:35:20 UTC] pqp <- lapply(readslist, FUN = function(x) { ppp <- plotQualityProfile(file.path(datadir, x)); ppp$facet$params$ncol <- 4; ppp })\n",
      "[19-07-2023 12:35:53 UTC] Saving quality profile plots to quality_Profile_R*.pdf\n",
      "[19-07-2023 12:35:57 UTC] out <- filterAndTrim(fwd=file.path(datadir,readslist$R1), filt=file.path(filt.dir,trimlist$R1),rev=file.path(datadir,readslist$R2), filt.rev=file.path(filt.dir,trimlist$R2),  maxEE=5, trimLeft=c(0, 0), truncQ=4, truncLen = c(0, 0), rm.phix=TRUE, compress=TRUE, verbose=TRUE, multithread=12, minLen=50)\n",
      "[19-07-2023 12:36:01 UTC] Creating output directory: /nephele_data/outputs/filtered_data\n",
      "[19-07-2023 12:36:03 UTC] reads.in reads.out\n",
      "[19-07-2023 12:36:03 UTC] 22831_S41_R1_subsample.fastq     25000     20511\n",
      "[19-07-2023 12:36:03 UTC] 22833_S45_R1_subsample.fastq     25000     20346\n",
      "[19-07-2023 12:36:03 UTC] 22349_S26_R1_subsample.fastq     25000     20929\n",
      "[19-07-2023 12:36:03 UTC] 22192_S22_R1_subsample.fastq     25000     21446\n",
      "[19-07-2023 12:36:03 UTC] 22187_S19_R1_subsample.fastq     25000     20753\n",
      "[19-07-2023 12:36:03 UTC] 22061_S5_R1_subsample.fastq      25000     20200\n",
      "[19-07-2023 12:36:03 UTC] 22057_S2_R1_subsample.fastq      25000     20969\n",
      "[19-07-2023 12:36:03 UTC] 22145_S14_R1_subsample.fastq     25000     18613\n",
      "[19-07-2023 12:36:03 UTC] 22350_S27_R1_subsample.fastq     25000     19778\n",
      "[19-07-2023 12:36:03 UTC] 23572_S307_R1_subsample.fastq    25000     17656\n",
      "[19-07-2023 12:36:03 UTC] Saved Vega-Lite data to: /nephele_data/outputs/readsInReadsOutVegaJSON.json\n",
      "[19-07-2023 12:36:03 UTC] Checking that trimmed files exist.\n",
      "[19-07-2023 12:36:04 UTC] list2env(checktrimfiles(A, filt.dir, trimlist), envir = environment())\n",
      "[19-07-2023 12:36:08 UTC] err <- lapply(trimlist, function(x) learnErrors(x, multithread=12, nbases=100000000,randomize=FALSE))\n",
      "[19-07-2023 12:36:14 UTC] 52511424 total bases in 201201 reads from 10 samples will be used for learning the error rates.\n",
      "[19-07-2023 12:37:36 UTC] 52366403 total bases in 201201 reads from 10 samples will be used for learning the error rates.\n",
      "[19-07-2023 12:39:31 UTC] pe <- lapply(err, function(x) plotErrors(x, nominalQ=TRUE))\n",
      "[19-07-2023 12:39:31 UTC] Saving 7 x 7 in image\n",
      "[19-07-2023 12:39:32 UTC] Warning: Transformation introduced infinite values in continuous y-axis\n",
      "[19-07-2023 12:39:32 UTC] Saving 7 x 7 in image\n",
      "[19-07-2023 12:39:33 UTC] Warning: Transformation introduced infinite values in continuous y-axis\n",
      "[19-07-2023 12:39:33 UTC] Saving 7 x 7 in image\n",
      "[19-07-2023 12:39:34 UTC] Warning: Transformation introduced infinite values in continuous y-axis\n",
      "[19-07-2023 12:39:34 UTC] Saving 7 x 7 in image\n",
      "[19-07-2023 12:39:35 UTC] Warning: Transformation introduced infinite values in continuous y-axis\n",
      "[19-07-2023 12:39:35 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:35 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22831_S41_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:35 UTC] Encountered 10011 unique sequences from 20511 total sequences read.\n",
      "[19-07-2023 12:39:36 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22831_S41_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:36 UTC] Encountered 16353 unique sequences from 20511 total sequences read.\n",
      "[19-07-2023 12:39:38 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:38 UTC] R1: 132 sequence variants were inferred from 10011 input unique sequences. R2: 106 sequence variants were inferred from 16353 input unique sequences.\n",
      "[19-07-2023 12:39:39 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:39 UTC] 14245 paired-reads (in 245 unique pairings) successfully merged out of 19342 (in 799 pairings) input.\n",
      "[19-07-2023 12:39:39 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:40 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22833_S45_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:40 UTC] Encountered 12613 unique sequences from 20346 total sequences read.\n",
      "[19-07-2023 12:39:40 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22833_S45_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:40 UTC] Encountered 18675 unique sequences from 20346 total sequences read.\n",
      "[19-07-2023 12:39:44 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:44 UTC] R1: 268 sequence variants were inferred from 12613 input unique sequences. R2: 82 sequence variants were inferred from 18675 input unique sequences.\n",
      "[19-07-2023 12:39:44 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:44 UTC] 10806 paired-reads (in 98 unique pairings) successfully merged out of 16476 (in 400 pairings) input.\n",
      "[19-07-2023 12:39:44 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:44 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22349_S26_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:44 UTC] Encountered 11655 unique sequences from 20929 total sequences read.\n",
      "[19-07-2023 12:39:45 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22349_S26_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:45 UTC] Encountered 17146 unique sequences from 20929 total sequences read.\n",
      "[19-07-2023 12:39:48 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:48 UTC] R1: 175 sequence variants were inferred from 11655 input unique sequences. R2: 95 sequence variants were inferred from 17146 input unique sequences.\n",
      "[19-07-2023 12:39:48 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:49 UTC] 12735 paired-reads (in 161 unique pairings) successfully merged out of 18253 (in 738 pairings) input.\n",
      "[19-07-2023 12:39:49 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:49 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22192_S22_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:49 UTC] Encountered 10687 unique sequences from 21446 total sequences read.\n",
      "[19-07-2023 12:39:49 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22192_S22_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:49 UTC] Encountered 16476 unique sequences from 21446 total sequences read.\n",
      "[19-07-2023 12:39:52 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:52 UTC] R1: 130 sequence variants were inferred from 10687 input unique sequences. R2: 95 sequence variants were inferred from 16476 input unique sequences.\n",
      "[19-07-2023 12:39:53 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:53 UTC] 17599 paired-reads (in 177 unique pairings) successfully merged out of 19561 (in 513 pairings) input.\n",
      "[19-07-2023 12:39:53 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:53 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22187_S19_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:53 UTC] Encountered 10100 unique sequences from 20753 total sequences read.\n",
      "[19-07-2023 12:39:53 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22187_S19_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:53 UTC] Encountered 16901 unique sequences from 20753 total sequences read.\n",
      "[19-07-2023 12:39:56 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:56 UTC] R1: 140 sequence variants were inferred from 10100 input unique sequences. R2: 99 sequence variants were inferred from 16901 input unique sequences.\n",
      "[19-07-2023 12:39:56 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:56 UTC] 17011 paired-reads (in 183 unique pairings) successfully merged out of 18987 (in 418 pairings) input.\n",
      "[19-07-2023 12:39:56 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:39:56 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22061_S5_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:56 UTC] Encountered 11283 unique sequences from 20200 total sequences read.\n",
      "[19-07-2023 12:39:57 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22061_S5_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:39:57 UTC] Encountered 17350 unique sequences from 20200 total sequences read.\n",
      "[19-07-2023 12:39:59 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:39:59 UTC] R1: 186 sequence variants were inferred from 11283 input unique sequences. R2: 83 sequence variants were inferred from 17350 input unique sequences.\n",
      "[19-07-2023 12:39:59 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:39:59 UTC] 14630 paired-reads (in 133 unique pairings) successfully merged out of 17632 (in 343 pairings) input.\n",
      "[19-07-2023 12:39:59 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:40:00 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22057_S2_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:00 UTC] Encountered 9268 unique sequences from 20969 total sequences read.\n",
      "[19-07-2023 12:40:00 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22057_S2_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:00 UTC] Encountered 15736 unique sequences from 20969 total sequences read.\n",
      "[19-07-2023 12:40:02 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:40:02 UTC] R1: 151 sequence variants were inferred from 9268 input unique sequences. R2: 148 sequence variants were inferred from 15736 input unique sequences.\n",
      "[19-07-2023 12:40:03 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:40:03 UTC] 16865 paired-reads (in 269 unique pairings) successfully merged out of 20052 (in 475 pairings) input.\n",
      "[19-07-2023 12:40:03 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:40:03 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22145_S14_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:03 UTC] Encountered 9540 unique sequences from 18613 total sequences read.\n",
      "[19-07-2023 12:40:03 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22145_S14_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:03 UTC] Encountered 16815 unique sequences from 18613 total sequences read.\n",
      "[19-07-2023 12:40:05 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:40:05 UTC] R1: 222 sequence variants were inferred from 9540 input unique sequences. R2: 79 sequence variants were inferred from 16815 input unique sequences.\n",
      "[19-07-2023 12:40:06 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:40:06 UTC] 12534 paired-reads (in 144 unique pairings) successfully merged out of 15578 (in 352 pairings) input.\n",
      "[19-07-2023 12:40:06 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:40:07 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22350_S27_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:07 UTC] Encountered 12092 unique sequences from 19778 total sequences read.\n",
      "[19-07-2023 12:40:07 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/22350_S27_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:07 UTC] Encountered 17357 unique sequences from 19778 total sequences read.\n",
      "[19-07-2023 12:40:10 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:40:10 UTC] R1: 214 sequence variants were inferred from 12092 input unique sequences. R2: 94 sequence variants were inferred from 17357 input unique sequences.\n",
      "[19-07-2023 12:40:10 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:40:10 UTC] 12376 paired-reads (in 128 unique pairings) successfully merged out of 16356 (in 470 pairings) input.\n",
      "[19-07-2023 12:40:10 UTC] derep <- lapply(trimlist, function(x) derepFastq(x[sample], verbose=TRUE))\n",
      "[19-07-2023 12:40:10 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/23572_S307_R1_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:10 UTC] Encountered 9005 unique sequences from 17656 total sequences read.\n",
      "[19-07-2023 12:40:11 UTC] Dereplicating sequence entries in Fastq file: /nephele_data/outputs/filtered_data/23572_S307_R2_subsample_trim.fastq.gz\n",
      "[19-07-2023 12:40:11 UTC] Encountered 16088 unique sequences from 17656 total sequences read.\n",
      "[19-07-2023 12:40:12 UTC] dd <- sapply(nameslist, function(x) dada(derep[[x]], err=err[[x]], multithread=12, verbose=F, priors = pseudo_priors[[x]]), USE.NAMES=TRUE, simplify=FALSE)\n",
      "[19-07-2023 12:40:12 UTC] R1: 162 sequence variants were inferred from 9005 input unique sequences. R2: 47 sequence variants were inferred from 16088 input unique sequences.\n",
      "[19-07-2023 12:40:13 UTC] mergePairs(dd$R1, derep$R1, dd$R2, derep$R2, verbose=TRUE, minOverlap=12, trimOverhang=FALSE, maxMismatch=0, justConcatenate=FALSE)\n",
      "[19-07-2023 12:40:13 UTC] 11701 paired-reads (in 85 unique pairings) successfully merged out of 15018 (in 267 pairings) input.\n",
      "[19-07-2023 12:40:13 UTC] seqtab <- makeSequenceTable(sampleVariants$sv)\n",
      "[19-07-2023 12:40:13 UTC] Removing sequences of length less than 75bp\n",
      "[19-07-2023 12:40:13 UTC] seqlengths <- nchar(colnames(seqtab))\n",
      "[19-07-2023 12:40:13 UTC] seqtab <- seqtab[,which(seqlengths >=75), drop=F]\n",
      "[19-07-2023 12:40:13 UTC] saveRDS(seqtab, file.path(interm.dir,\"seqtab_min75.rds\"))\n",
      "[19-07-2023 12:40:13 UTC] seqtabnochimera <- removeBimeraDenovo(seqtab, verbose=TRUE, multithread=12)\n",
      "[19-07-2023 12:40:13 UTC] Identified 568 bimeras out of 1349 input sequences.\n",
      "[19-07-2023 12:40:13 UTC] % Reads remaining after chimera removal: 71.4808330130532\n",
      "[19-07-2023 12:40:13 UTC] seqtab <- seqtabnochimera\n",
      "[19-07-2023 12:40:13 UTC] Track Reads\n",
      "             denoisedF denoisedR merged filter75 nochim\n",
      "[19-07-2023 12:40:13 UTC] A22831           20093     19494  14245    14245   7921\n",
      "[19-07-2023 12:40:13 UTC] A22833           18688     16924  10806    10806  10054\n",
      "[19-07-2023 12:40:13 UTC] A22349           19980     18528  12735    12735   8165\n",
      "[19-07-2023 12:40:13 UTC] A22192           20688     19840  17599    17599  13051\n",
      "[19-07-2023 12:40:13 UTC] A22187           20173     19243  17011    17011  10509\n",
      "[19-07-2023 12:40:13 UTC] A22061           19115     17917  14630    14630  12260\n",
      "[19-07-2023 12:40:13 UTC] A22057           20673     20147  16865    16865   9998\n",
      "[19-07-2023 12:40:13 UTC] A22145           17757     15789  12534    12534   9239\n",
      "[19-07-2023 12:40:13 UTC] A22350           18350     16709  12376    12376   9838\n",
      "[19-07-2023 12:40:13 UTC] 7pRecSw478.1     16806     15202  11701    11701   9397\n",
      "[19-07-2023 12:40:13 UTC] Saved Vega-Lite data to: /nephele_data/outputs/trackReadsVegaJSON.json\n",
      "[19-07-2023 12:40:13 UTC] rep_seq_names <- dada2fasta(seqtab, filename=\"/nephele_data/outputs/seq.fasta\")\n",
      "[19-07-2023 12:40:13 UTC] rep_seq_names <- make_seq_names(seqtab, nametype)\n",
      "[19-07-2023 12:40:13 UTC] writeFasta(seqs, file=\"/nephele_data/outputs/seq.fasta\")\n",
      "[19-07-2023 12:40:13 UTC] Taxonomic assignment with rdp\n",
      "[19-07-2023 12:40:17 UTC] taxa <- assignTaxonomy(seqtab, refdb, multithread=12, minBoot=80, tryRC=TRUE, verbose=TRUE)\n",
      "[19-07-2023 12:42:09 UTC] Finished processing reference fasta.\n",
      "[19-07-2023 12:42:09 UTC] Species assignment with dada2::addSpecies\n",
      "[19-07-2023 12:42:09 UTC] taxa.genus <- taxa; rm(taxa);\n",
      "[19-07-2023 12:42:13 UTC] taxa <- addSpecies(taxa.genus, refdb_species, verbose=TRUE, tryRC=TRUE, n=4000, allowMultiple =FALSE)\n",
      "[19-07-2023 12:43:28 UTC] 3 out of 781 were assigned to the species level.\n",
      "[19-07-2023 12:43:28 UTC] Of which 2 had genera consistent with the input table.Garbage collection 247 = 164+40+43 (level 2) ...\n",
      "[19-07-2023 12:43:28 UTC] 432.3 Mbytes of cons cells used (56%)\n",
      "[19-07-2023 12:43:28 UTC] 175.0 Mbytes of vectors used (5%)\n",
      "[19-07-2023 12:43:28 UTC] rep_seq_names <- make_seq_names(seqtab, nametype)\n",
      "[19-07-2023 12:43:28 UTC] writeFasta(seqs, file=\"/nephele_data/outputs/seq.fasta\")\n",
      "[19-07-2023 12:43:28 UTC] colnames(seqtab) <- replace_names(colnames(seqtab), rep_seq_names)\n",
      "[19-07-2023 12:43:28 UTC] row.names(taxtab) <- replace_names(row.names(taxtab), rep_seq_names)\n",
      "[19-07-2023 12:43:30 UTC] write_biom(dada2biom(seqtab,taxtab, metadata = metadata), file.path(outdir, \"taxa.biom\"))\n",
      "[19-07-2023 12:43:30 UTC] dada2text(seqtab, taxtab, file.path(outdir, \"OTU_table.txt\"))\n",
      "[19-07-2023 12:43:30 UTC] dada2taxonomy(taxtab, file.path(outdir, \"taxonomy_table.txt\"))\n",
      "[19-07-2023 12:43:30 UTC] Garbage collection 248 = 164+40+44 (level 2) ...\n",
      "[19-07-2023 12:43:30 UTC] 432.1 Mbytes of cons cells used (56%)\n",
      "[19-07-2023 12:43:31 UTC] 111.6 Mbytes of vectors used (4%)\n",
      "[19-07-2023 12:43:31 UTC] Summarizing biom file to /nephele_data/outputs/otu_summary_table.txt.\n",
      "[19-07-2023 12:43:31 UTC] Creating phylogenetic trees\n",
      "[19-07-2023 12:43:33 UTC] Running command: mafft --preservecase --inputorder --thread 12 /nephele_data/outputs/seq.fasta > /nephele_data/outputs/phylo/aligned_seq.fasta\n",
      "[19-07-2023 12:43:37 UTC] Running command: FastTreeMP -quote -nt /nephele_data/outputs/phylo/aligned_seq.fasta > /nephele_data/outputs/phylo/unrooted_tree.nwk\n",
      "[19-07-2023 12:43:39 UTC] Finished creating trees: /nephele_data/outputs/phylo/rooted_tree.nwk, /nephele_data/outputs/phylo/unrooted_tree.nwk\n",
      "[19-07-2023 12:43:39 UTC] Checking output file from dada2 pipeline required by data visualization pipeline.\n",
      "[19-07-2023 12:43:39 UTC] Running data visualization pipeline\n",
      "[19-07-2023 12:43:43 UTC] Running with args: {'datafile': '/nephele_data/outputs/OTU_table.txt', 'outdir': '/nephele_data/outputs/', 'logfilename': '/var/log/job.log', 'sampdepth': 10054, 'mapfile': '/nephele_data/outputs/N2_16S_example_mapping_file_3_corrected.txt.no_gz', 'tsvfile': True}\n",
      "[19-07-2023 12:43:44 UTC] R version 4.3.1 (2023-06-16)\n",
      "[19-07-2023 12:43:44 UTC] Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "[19-07-2023 12:43:44 UTC] Running under: Debian GNU/Linux 11 (bullseye)\n",
      "[19-07-2023 12:43:44 UTC] Matrix products: default\n",
      "[19-07-2023 12:43:44 UTC] BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0\n",
      "[19-07-2023 12:43:44 UTC] LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0\n",
      "[19-07-2023 12:43:44 UTC] locale:\n",
      " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
      " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
      " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C\n",
      "[19-07-2023 12:43:44 UTC] [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C\n",
      "[19-07-2023 12:43:44 UTC] time zone: America/New_York\n",
      "[19-07-2023 12:43:44 UTC] tzcode source: system (glibc)\n",
      "[19-07-2023 12:43:44 UTC] attached base packages:\n",
      "[19-07-2023 12:43:44 UTC] [1] stats     graphics  grDevices utils     datasets  methods   base\n",
      "[19-07-2023 12:43:44 UTC] other attached packages:\n",
      " [1] htmlwidgets_1.6.2 jsonlite_1.8.7    plotly_4.10.2     ampvis2_2.7.4    \n",
      " [5] ggplot2_3.4.2     vegan_2.6-4       lattice_0.20-45   permute_0.9-7    \n",
      " [9] htmltools_0.5.5   morpheus_0.1.1.1\n",
      "[19-07-2023 12:43:44 UTC] loaded via a namespace (and not attached):\n",
      " [1] Matrix_1.5-3       gtable_0.3.3       crayon_1.5.2       dplyr_1.1.2       \n",
      " [5] compiler_4.3.1     tidyselect_1.2.0   Rcpp_1.0.11        stringr_1.5.0     \n",
      " [9] parallel_4.3.1     tidyr_1.3.0        cluster_2.1.3      splines_4.3.1\n",
      "[19-07-2023 12:43:44 UTC] [13] scales_1.2.1       fastmap_1.1.1      plyr_1.8.8         R6_2.5.1\n",
      "[19-07-2023 12:43:44 UTC] [17] generics_0.1.3     MASS_7.3-58.3      ggrepel_0.9.3      tibble_3.2.1\n",
      "[19-07-2023 12:43:44 UTC] [21] munsell_0.5.0      pillar_1.9.0       RColorBrewer_1.1-3 rlang_1.1.1\n",
      "[19-07-2023 12:43:44 UTC] [25] utf8_1.2.3         stringi_1.7.12     lazyeval_0.2.2     viridisLite_0.4.2\n",
      "[19-07-2023 12:43:44 UTC] [29] cli_3.6.1          withr_2.5.0        magrittr_2.0.3     mgcv_1.8-42\n",
      "[19-07-2023 12:43:44 UTC] [33] digest_0.6.33      grid_4.3.1         lifecycle_1.0.3    nlme_3.1-162\n",
      "[19-07-2023 12:43:44 UTC] [37] vctrs_0.6.3        data.table_1.14.8  glue_1.6.2         ape_5.7-1\n",
      "[19-07-2023 12:43:44 UTC] [41] fansi_1.0.4        colorspace_2.1-0   purrr_1.0.1        httr_1.4.6\n",
      "[19-07-2023 12:43:44 UTC] [45] tools_4.3.1        pkgconfig_2.0.3\n",
      "[19-07-2023 12:43:44 UTC] \"allgraphs\"(datafile=\"/nephele_data/outputs/OTU_table.txt\", outdir=\"/nephele_data/outputs//graphs\", mapfile=\"/nephele_data/outputs/N2_16S_example_mapping_file_3_corrected.txt.no_gz\",tsvfile=TRUE, ...)\n",
      "[19-07-2023 12:43:44 UTC] Reading in map file /nephele_data/outputs/N2_16S_example_mapping_file_3_corrected.txt.no_gz\n",
      "[19-07-2023 12:43:44 UTC] Reading in OTU file /nephele_data/outputs/OTU_table.txt\n",
      "[19-07-2023 12:43:44 UTC] otu <- read.delim(datafile, check.names = FALSE, na.strings = '', row.names = 1)\n",
      "[19-07-2023 12:43:44 UTC] tax <- otu[,!names(otu) %in% map$SampleID]\n",
      "[19-07-2023 12:43:44 UTC] otu <- otu[, names(otu) %in% map$SampleID, drop=F]\n",
      "[19-07-2023 12:43:44 UTC] otu <- cbind(otu, tax)\n",
      "[19-07-2023 12:43:44 UTC] amp <- amp_load(otu, map)\n",
      "[19-07-2023 12:43:45 UTC] Warning: Could not find a column named OTU/ASV in otutable, using rownames as sample ID's\n",
      "[19-07-2023 12:43:45 UTC] ampvis2 object with 3 elements.\n",
      "[19-07-2023 12:43:45 UTC] Summary of OTU table:\n",
      "     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads \n",
      "          10          781       100432         7921        13051         9918 \n",
      "   Avg#Reads \n",
      "     10043.2\n",
      "[19-07-2023 12:43:45 UTC] Assigned taxonomy:\n",
      "    Kingdom      Phylum       Class       Order      Family       Genus \n",
      "  781(100%) 771(98.72%) 770(98.59%) 760(97.31%) 702(89.88%) 527(67.48%) \n",
      "    Species \n",
      "   2(0.26%)\n",
      "[19-07-2023 12:43:45 UTC] Metadata variables: 7 \n",
      " SampleID, ForwardFastqFile, ReverseFastqFile, TreatmentGroup, Animal, Day, Description\n",
      "[19-07-2023 12:43:45 UTC] Rarefaction curve\n",
      "[19-07-2023 12:43:45 UTC] rarefactioncurve(outdir = outdir, amp = amp, colors = allcols, pipeline=TRUE)\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 8\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 9\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 4\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 4\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 24\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 9\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 6\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 7\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 2\n",
      "[19-07-2023 12:43:45 UTC] Warning in vegan::rarefy(abund[i, ], n) :\n",
      "  most observed count data have counts 1, but smallest count is 30\n",
      "[19-07-2023 12:43:45 UTC] Saving plot to /nephele_data/outputs/graphs/rarecurve.html\n",
      "[19-07-2023 12:43:46 UTC] Warning in plotly::config(pp, cloud = T, edits = list(titleText = T, legendText = T,  :\n",
      "  The `cloud` argument is deprecated. Use `showSendToCloud` instead.\n",
      "[19-07-2023 12:43:46 UTC] Saving rarefaction curve table to /nephele_data/outputs//graphs/rarecurve.txt\n",
      "[19-07-2023 12:43:46 UTC] Relative abundance heatmaps\n",
      "[19-07-2023 12:43:46 UTC] morphheatmap(outdir = outdir, amp = amp, colors=allcols, filter_level = 5)\n",
      "[19-07-2023 12:43:46 UTC] Filter taxa below 5 counts/abundance.\n",
      "[19-07-2023 12:43:46 UTC] amp <- filterlowabund(amp, level = 5, abs=T)\n",
      "[19-07-2023 12:43:46 UTC] Calculate relative abundance.\n",
      "[19-07-2023 12:43:46 UTC] amp <- subsetamp(amp, sampdepth = NULL, rarefy=FALSE, normalise = TRUE, printsummary = FALSE)\n",
      "[19-07-2023 12:43:46 UTC] 0 samples have been filtered.\n",
      "[19-07-2023 12:43:46 UTC] makeheatmap(\"seq\", amp)\n",
      "[19-07-2023 12:43:48 UTC] heatmap <- morpheus(mat, columns=columns, columnAnnotations = amptax$metadata, columnColorModel = list(type=as.list(colors)), colorScheme = list(scalingMode=\"fixed\", values=values, colors=hmapcolors, stepped=FALSE), rowAnnotations = amptax$tax, rows = rows, dendrogram=\"none\")\n",
      "[19-07-2023 12:43:48 UTC] Saving plot to /nephele_data/outputs/graphs/seq_heatmap.html\n",
      "[19-07-2023 12:43:48 UTC] Sampling depth: 10054\n",
      "[19-07-2023 12:43:48 UTC] Filter samples below 10054 counts.\n",
      "[19-07-2023 12:43:48 UTC] amp <- amp_subset_samples(amp, minreads = 10054, ...)\n",
      "[19-07-2023 12:43:48 UTC] 6 samples and 448 OTUs have been filtered\n",
      "[19-07-2023 12:43:48 UTC] Before: 10 samples and 781 OTUs\n",
      "[19-07-2023 12:43:48 UTC] After: 4 samples and 333 OTUs\n",
      "[19-07-2023 12:43:48 UTC] Saving excluded sample ids to /nephele_data/outputs//graphs/samples_being_ignored.txt\n",
      "[19-07-2023 12:43:48 UTC] ampvis2 object with 3 elements.\n",
      "[19-07-2023 12:43:48 UTC] Summary of OTU table:\n",
      "     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads \n",
      "           4          333        45874        10054        13051      11384.5 \n",
      "   Avg#Reads \n",
      "     11468.5\n",
      "[19-07-2023 12:43:48 UTC] Assigned taxonomy:\n",
      "    Kingdom      Phylum       Class       Order      Family       Genus \n",
      "  333(100%)  330(99.1%)  330(99.1%)  320(96.1%) 296(88.89%) 227(68.17%) \n",
      "    Species \n",
      "    2(0.6%)\n",
      "[19-07-2023 12:43:48 UTC] Metadata variables: 7 \n",
      " SampleID, ForwardFastqFile, ReverseFastqFile, TreatmentGroup, Animal, Day, Description\n",
      "[19-07-2023 12:43:48 UTC] PCoA plot with binomial distance\n",
      "[19-07-2023 12:43:48 UTC] pcoaplot(outdir = outdir, amp = ampsub, distm = \"binomial\", colors = allcols)\n",
      "[19-07-2023 12:43:49 UTC] pcoa <- amp_ordinate(amp, filter_species =0.1,type=\"PCOA\", distmeasure =\"binomial\",sample_color_by = \"TreatmentGroup\", detailed_output = TRUE, transform=\"none\")\n",
      "[19-07-2023 12:43:49 UTC] Saving plot to /nephele_data/outputs/graphs/pcoa_binomial.html\n",
      "[19-07-2023 12:43:49 UTC] Warning in plotly::config(pp, cloud = T, edits = list(titleText = T, legendText = T,  :\n",
      "  The `cloud` argument is deprecated. Use `showSendToCloud` instead.\n",
      "[19-07-2023 12:43:49 UTC] Saving binomial PCoA table to /nephele_data/outputs//graphs/pcoa_binomial.txt\n",
      "[19-07-2023 12:43:49 UTC] Making top species table.\n",
      "[19-07-2023 12:43:49 UTC] Saving table to /nephele_data/outputs//graphs/top_85_species_table.txt\n",
      "[19-07-2023 12:43:49 UTC] Rarefying OTU Table to  10054 reads.\n",
      "[19-07-2023 12:43:49 UTC] set.seed(500)\n",
      "[19-07-2023 12:43:49 UTC] otu <- rrarefy(t(amp$abund), sampdepth)\n",
      "[19-07-2023 12:43:49 UTC] Warning in rrarefy(t(amp$abund), sampdepth) :\n",
      "  function should be used for observed counts, but smallest count is 4\n",
      "[19-07-2023 12:43:49 UTC] amp <- amp_subset_samples(amp, minreads = 10054, ...)\n",
      "[19-07-2023 12:43:49 UTC] 0 samples have been filtered.\n",
      "[19-07-2023 12:43:49 UTC] ampvis2 object with 3 elements.\n",
      "[19-07-2023 12:43:49 UTC] Summary of OTU table:\n",
      "     Samples         OTUs  Total#Reads    Min#Reads    Max#Reads Median#Reads \n",
      "           4          333        40216        10054        10054        10054 \n",
      "   Avg#Reads \n",
      "       10054\n",
      "[19-07-2023 12:43:49 UTC] Assigned taxonomy:\n",
      "    Kingdom      Phylum       Class       Order      Family       Genus \n",
      "  333(100%)  330(99.1%)  330(99.1%)  320(96.1%) 296(88.89%) 227(68.17%) \n",
      "    Species \n",
      "    2(0.6%)\n",
      "[19-07-2023 12:43:49 UTC] Metadata variables: 7 \n",
      " SampleID, ForwardFastqFile, ReverseFastqFile, TreatmentGroup, Animal, Day, Description\n",
      "[19-07-2023 12:43:49 UTC] Saving rarefied OTU Table to  /nephele_data/outputs//graphs/rarefied_OTU_table_10054.txt\n",
      "[19-07-2023 12:43:49 UTC] Making heatmap from rarefied counts.\n",
      "[19-07-2023 12:43:49 UTC] morphheatmap(outdir = outdir, amp = amprare, colors=allcols, filter_level = 5, filesuffix = \"_rarefied\")\n",
      "[19-07-2023 12:43:49 UTC] Filter taxa below 5 counts/abundance.\n",
      "[19-07-2023 12:43:49 UTC] amp <- filterlowabund(amp, level = 5, abs=T)\n",
      "[19-07-2023 12:43:49 UTC] Calculate relative abundance.\n",
      "[19-07-2023 12:43:49 UTC] amp <- subsetamp(amp, sampdepth = NULL, rarefy=FALSE, normalise = TRUE, printsummary = FALSE)\n",
      "[19-07-2023 12:43:49 UTC] 0 samples have been filtered.\n",
      "[19-07-2023 12:43:49 UTC] makeheatmap(\"seq\", amp)\n",
      "[19-07-2023 12:43:49 UTC] heatmap <- morpheus(mat, columns=columns, columnAnnotations = amptax$metadata, columnColorModel = list(type=as.list(colors)), colorScheme = list(scalingMode=\"fixed\", values=values, colors=hmapcolors, stepped=FALSE), rowAnnotations = amptax$tax, rows = rows, dendrogram=\"none\")\n",
      "[19-07-2023 12:43:49 UTC] Saving plot to /nephele_data/outputs/graphs/seq_heatmap_rarefied.html\n",
      "[19-07-2023 12:43:49 UTC] Normalizing rarefied OTU table to 100 for Bray-Curtis distance.\n",
      "[19-07-2023 12:43:49 UTC] 0 samples have been filtered.\n",
      "[19-07-2023 12:43:49 UTC] pcoaplot(outdir = outdir, amp = ampbc, distm = \"bray\", colors = allcols, filesuffix=\"_rarefied\")\n",
      "[19-07-2023 12:43:49 UTC] pcoa <- amp_ordinate(amp, filter_species =0.1,type=\"PCOA\", distmeasure =\"bray\",sample_color_by = \"TreatmentGroup\", detailed_output = TRUE, transform=\"none\")\n",
      "[19-07-2023 12:43:49 UTC] Saving plot to /nephele_data/outputs/graphs/pcoa_bray_rarefied.html\n",
      "[19-07-2023 12:43:49 UTC] Warning in plotly::config(pp, cloud = T, edits = list(titleText = T, legendText = T,  :\n",
      "  The `cloud` argument is deprecated. Use `showSendToCloud` instead.\n",
      "[19-07-2023 12:43:49 UTC] Saving bray PCoA table to /nephele_data/outputs//graphs/pcoa_bray_rarefied.txt\n",
      "[19-07-2023 12:43:49 UTC] Alpha diversity boxplot\n",
      "[19-07-2023 12:43:49 UTC] adivboxplot(outdir = outdir, amp = amprare, sampdepth = sampdepth, colors = allcols, pipeline=TRUE)\n",
      "[19-07-2023 12:43:49 UTC] alphadiv <- amp_alphadiv(amp, measure=\"shannon\", richness = TRUE, rarefy = 10054)\n",
      "[19-07-2023 12:43:49 UTC] Warning: The data you have provided does not have\n",
      "[19-07-2023 12:43:49 UTC] any singletons. This is highly suspicious. Results of richness\n",
      "[19-07-2023 12:43:49 UTC] estimates (for example) are probably unreliable, or wrong, if you have already\n",
      "[19-07-2023 12:43:49 UTC] trimmed low-abundance taxa from the data.\n",
      "[19-07-2023 12:43:49 UTC] We recommend that you find the un-trimmed data and retry.\n",
      "[19-07-2023 12:43:50 UTC] Saving alpha diversity table to /nephele_data/outputs//graphs/alphadiv.txt\n",
      "[19-07-2023 12:43:50 UTC] Saving plot to /nephele_data/outputs/graphs/alphadiv.html\n",
      "[19-07-2023 12:43:50 UTC] Warning in plotly::config(pp, cloud = T, edits = list(titleText = T, legendText = T,  :\n",
      "  The `cloud` argument is deprecated. Use `showSendToCloud` instead.\n",
      "[19-07-2023 12:43:51 UTC] \"allgraphs\" complete.\n",
      "[19-07-2023 12:43:51 UTC] Result 'ref_db' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'error_rate_r1' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'rooted_tree' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'top_species_table' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'track_reads' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'sampling_depth' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'otu_summary_table' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'biom' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'alphadiv' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'pcoa_binomial' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'rarecurve' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'logfile_debug' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'quality_profile_r2' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'error_rate_r2' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'pcoa_bray' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'otu_table' reported\n",
      "[19-07-2023 12:43:51 UTC] Optional result 'species_heatmap' does not exist: /nephele_data/outputs/graphs/Species_heatmap.html\n",
      "[19-07-2023 12:43:51 UTC] Result 'rarefied_otu_table' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'seq_fasta' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'quality_profile_r1' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'taxonomy_table' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'reads_in_reads_out' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'seq_heatmap' reported\n",
      "[19-07-2023 12:43:51 UTC] Result 'taxmethod' reported\n",
      "[19-07-2023 12:43:52 UTC] Results tarball does not exist: /mnt/EFS/user_uploads/b73de8bfdd22_reported_results.tar.gz. Creating.\n",
      "[19-07-2023 12:43:52 UTC] Created results tarball: /mnt/EFS/user_uploads/b73de8bfdd22_reported_results.tar.gz\n",
      "[19-07-2023 12:43:52 UTC] Uploaded to S3: /mnt/EFS/user_uploads/b73de8bfdd22_reported_results.tar.gz\n",
      "[19-07-2023 12:43:52 UTC] Uploaded to S3: /mnt/EFS/user_uploads/b73de8bfdd22/outputs/b73de8bfdd22_results_registry.json\n",
      "[19-07-2023 12:43:52 UTC] DADA2 pipeline complete.\n",
      "[19-07-2023 12:43:56 UTC] None\n"
     ]
    }
   ],
   "source": [
    "formatted_logs, token = lw.return_formatted_logs()\n",
    "\n",
    "print(formatted_logs)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4d7f7e559600332ff0dc37dc6eb58bfc6431dddd5a32651386b47ac3a3367ecf"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
